--- a/src/Game/Effect/MultiEmitter.cpp
+++ b/src/Game/Effect/MultiEmitter.cpp
@@ -4,6 +4,7 @@
 #include "Game/Effect/MultiEmitterCallBack.hpp"
 #include "Game/Effect/MultiEmitterParticleCallBack.hpp"
 #include "Game/Effect/SingleEmitter.hpp"
+#include "Game/Effect/SyncBckEffectInfo.hpp"
 #include "Game/Util/HashUtil.hpp"
 #include "Game/Util/MemoryUtil.hpp"
 #include "Game/Util/StringUtil.hpp"
@@ -151,8 +152,9 @@
     }
 }
 
-/* void MultiEmitter::scanParticleEmitter(EffectSystem* pSystem) {
-} */
+void MultiEmitter::scanParticleEmitter(EffectSystem* pSystem) {
+    std::for_each_array(mEmitters.begin(), mEmitters.end(), std::bind2nd(std::mem_func(&SingleEmitter::scanParticleEmitter), pSystem));
+}
 
 void MultiEmitter::forceFollowOn() {
     mCallBack->forceFollowOn();
@@ -166,19 +168,23 @@
     mCallBack->forceScaleOn();
 }
 
-/*
-void MultiEmitter::initSyncBck(XanimePlayer* pPlayer, const char* pName, s32 i1, f32 f2) {
-}
-
-void MultiEmitter::onDeleteSyncBck(bool, f32) {
-}
-
-void MultiEmitter::addSyncBck(const XanimePlayer*, const char*) {
-}
-
-void MultiEmitter::setContinueBckEnd(bool) {
-}
- */
+void MultiEmitter::initSyncBck(XanimePlayer* pPlayer, const char* pName, s32 count, f32 startFrame) {
+    _24 = new SyncBckEffectInfo(pPlayer, pName, count, 0.0f, -1.0f, false);
+    _24->mStartFrame = startFrame;
+}
+
+void MultiEmitter::onDeleteSyncBck(bool, f32 endFrame) {
+    _24->mEndFrame = endFrame;
+}
+
+void MultiEmitter::addSyncBck(const XanimePlayer* pPlayer, const char* pName) {
+    _24->addBck(pPlayer, pName);
+}
+
+void MultiEmitter::setContinueBckEnd(bool continueBckEnd) {
+    _24->mContinueBckEnd = continueBckEnd;
+}
+
 void MultiEmitter::onCreateSyncClipping() {
     turnFlagOn(JPAEmtrStts_FirstEmit);
 }
--- /dev/null
+++ b/src/Game/Effect/SyncBckEffectInfo.cpp
@@ -0,0 +1,66 @@
+#include "Game/Effect/SyncBckEffectInfo.hpp"
+#include "Game/Animation/XanimePlayer.hpp"
+#include "Game/Animation/XanimeResource.hpp"
+#include "Game/Util/StringUtil.hpp"
+#include <JSystem/J3DGraphAnimator/J3DAnimation.hpp>
+
+SyncBckEffectInfo::BckResourceInfo::BckResourceInfo(const XanimePlayer* pPlayer, const char* pName) : mName(pName), mResource(nullptr) {
+    mResource = static_cast< J3DAnmTransform* >(pPlayer->mResourceTable->findResMotion(pName));
+}
+
+bool SyncBckEffectInfo::BckResourceInfo::isLoop() const {
+    return mResource->getAttribute() == 2 || mResource->getAttribute() == 4;
+}
+
+SyncBckEffectInfo::SyncBckEffectInfo(const XanimePlayer* pPlayer, const char* pName, s32 count, f32 startFrame, f32 endFrame, bool continueBckEnd)
+    : mBckResources(), mStartFrame(startFrame), mEndFrame(endFrame), mContinueBckEnd(continueBckEnd) {
+    mBckResources.init(count);
+    addBck(pPlayer, pName);
+}
+
+void SyncBckEffectInfo::addBck(const XanimePlayer* pPlayer, const char* pName) {
+    mBckResources.push_back(new BckResourceInfo(pPlayer, pName));
+}
+
+bool SyncBckEffectInfo::isRegisteredBck(const char* pName) const {
+    if (pName == nullptr) {
+        return false;
+    }
+
+    for (BckResourceInfo* const* pInfo = mBckResources.begin(); pInfo != mBckResources.end(); pInfo++) {
+        if (MR::isEqualStringCase(pName, (*pInfo)->mName)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+bool SyncBckEffectInfo::isBckLoop(const char* pName) const {
+    if (pName == nullptr) {
+        return false;
+    }
+
+    BckResourceInfo* pResource = nullptr;
+    for (BckResourceInfo* const* pInfo = mBckResources.begin(); pInfo != mBckResources.end(); pInfo++) {
+        if (MR::isEqualStringCase((*pInfo)->mName, pName)) {
+            pResource = *pInfo;
+            break;
+        }
+    }
+
+    if (pResource == nullptr) {
+        return false;
+    }
+
+    return pResource->isLoop();
+}
+
+namespace MR {
+    namespace Effect {
+        bool isExistSyncBckDeleteFrame(const SyncBckEffectInfo* pInfo) {
+            f32 endFrame = pInfo->mEndFrame;
+            return 0.0f <= endFrame;
+        }
+    }
+}
--- /dev/null
+++ b/include/Game/Effect/SyncBckEffectInfo.hpp
@@ -0,0 +1,36 @@
+#pragma once
+
+#include "Game/Util/Array.hpp"
+
+class J3DAnmTransform;
+class XanimePlayer;
+
+class SyncBckEffectInfo {
+public:
+    class BckResourceInfo {
+    public:
+        BckResourceInfo(const XanimePlayer*, const char*);
+
+        bool isLoop() const;
+
+        /* 0x00 */ const char* mName;
+        /* 0x04 */ J3DAnmTransform* mResource;
+    };
+
+    SyncBckEffectInfo(const XanimePlayer*, const char*, s32, f32, f32, bool);
+
+    void addBck(const XanimePlayer*, const char*);
+    bool isRegisteredBck(const char*) const;
+    bool isBckLoop(const char*) const;
+
+    /* 0x00 */ MR::Vector< MR::AssignableArray< BckResourceInfo* > > mBckResources;
+    /* 0x0C */ f32 mStartFrame;
+    /* 0x10 */ f32 mEndFrame;
+    /* 0x14 */ bool mContinueBckEnd;
+};
+
+namespace MR {
+    namespace Effect {
+        bool isExistSyncBckDeleteFrame(const SyncBckEffectInfo*);
+    }
+}
