--- a/src/Game/Player/J3DModelX.cpp
+++ b/src/Game/Player/J3DModelX.cpp
@@ -18,26 +18,12 @@
 #include <revolution/gd/GDBase.h>
 #include <revolution/gx.h>
 
-extern "C" {
-void GDSetCullMode(int);
-void GDSetAlphaCompare(int, int, int, int, int);
-void GDSetZMode(int, int, int);
-void GDSetGenMode2(int, int, int, int, int);
-void GDSetTevDirect(int);
-void GDSetChanCtrl(int, int, int, int, int, int, int);
-void GDSetBlendModeEtc(int, int, int, int, int, int, int);
-void GDSetDstAlpha(int, int);
-void GDSetTevAlphaCalcAndSwap(int, int, int, int, int, int, int, int, int, int, int, int);
-void GDSetTevOrder(int, int, int, int, int, int, int);
-void GDSetTevColorCalc(int, int, int, int, int, int, int, int, int, int);
-void GDSetTevColor(int, GXColor);
-void GDSetTexLookupMode(int, int, int, int, int, f32, f32, f32, int, int, int);
-void GDSetTexImgAttr(int, int, int, int);
-void GDSetTexImgPtr(int, void*);
-void GDSetTexCoordGen(int, int, int, int, int);
-void GDSetChanAmbColor(int, GXColor);
-void GDSetChanMatColor(int, GXColor);
-}
+#include <revolution/gd/GDGeometry.h>
+#include <revolution/gd/GDTev.h>
+#include <revolution/gd/GDPixel.h>
+#include <revolution/gd/GDIndirect.h>
+#include <revolution/gd/GDLight.h>
+#include <revolution/gd/GDTexture.h>
 
 class J3DMtxBuffer2 : public J3DMtxBuffer {
 public:
@@ -259,30 +245,30 @@
     }
 
     if (_1B8 != nullptr) {
-        _1B4 = reinterpret_cast< u32 >(_1B8);
+        _1B4 = _1B8;
         _1B8 = nullptr;
     }
 
     if (_1B4 != 0 && _1BC != 0) {
-        GXCallDisplayList(reinterpret_cast< void* >(_1B4), _1BC);
+        GXCallDisplayList(_1B4, _1BC);
     }
 
     if (_1C8[_1C0] != nullptr) {
-        _1C4[_1C0] = reinterpret_cast< u32 >(_1C8[_1C0]);
+        _1C4[_1C0] = _1C8[_1C0];
         _1C8[_1C0] = nullptr;
     }
 
     if (_1C4[_1C0] != 0 && _1CC[_1C0] != 0) {
-        GXCallDisplayList(reinterpret_cast< void* >(_1C4[_1C0]), _1CC[_1C0]);
+        GXCallDisplayList(_1C4[_1C0], _1CC[_1C0]);
     }
 
     if (mMaterialCallback != nullptr) {
-        mMaterialCallback(reinterpret_cast< void* >(_128), _1C0);
+        mMaterialCallback(_128, _1C0);
     }
 
     J3DShape* originalShape = nullptr;
     if (_12C != 0) {
-        J3DModel* shapeModel = reinterpret_cast< J3DModel* >(_12C);
+        J3DModel* shapeModel = _12C;
         originalShape = shapePacket->mpShape;
         J3DShape* replacementShape = shapeModel->mShapePacket[material->mShape->mIndex].mpShape;
         shapePacket->mpShape = replacementShape;
@@ -348,140 +334,140 @@
     GDInitGDLObj(&displayList, displayListBuffer, sizeof(displayListBuffer));
     __GDCurrentDL = &displayList;
 
-    GDSetCullMode(1);
+    GDSetCullMode(static_cast< GXCullMode >(1));
     storeDisplayList(&displayList, 0);
 
-    GDSetAlphaCompare(7, 0, 1, 7, 0);
-    GDSetZMode(0, 7, 0);
-    GDSetGenMode2(0, 1, 1, 0, 2);
-    GDSetTevDirect(0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetBlendModeEtc(0, 0, 5, 0, 0, 1, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(1), static_cast< GXCompare >(7), 0);
+    GDSetZMode(0, static_cast< GXCompare >(7), 0);
+    GDSetGenMode2(0, 1, 1, 0, static_cast< GXCullMode >(2));
+    GDSetTevDirect(static_cast< GXTevStageID >(0));
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(0), static_cast< GXBlendFactor >(0), static_cast< GXBlendFactor >(5), static_cast< GXLogicOp >(0), 0, 1, 0);
     GDSetDstAlpha(1, 0xFF);
-    GDSetTevAlphaCalcAndSwap(0, 7, 7, 7, 7, 0, 1, 3, 1, 0, 0, 0);
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(1), static_cast< GXTevScale >(3), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 1);
 
-    GDSetCullMode(1);
-    GDSetZMode(1, 3, 1);
-    GDSetBlendModeEtc(1, 4, 1, 0, 1, 0, 0);
-    GDSetGenMode2(1, 0, 1, 0, 1);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetChanCtrl(5, 0, 0, 0, 0, 2, 2);
-    GDSetTevOrder(0, 0, 0, 0xFF, 1, 1, 0xFF);
-    GDSetTevColorCalc(0, 8, 0xF, 0xF, 0xF, 0, 0, 0, 1, 0);
-    GDSetTevAlphaCalcAndSwap(0, 7, 7, 7, 7, 0, 1, 3, 1, 0, 0, 0);
+    GDSetCullMode(static_cast< GXCullMode >(1));
+    GDSetZMode(1, static_cast< GXCompare >(3), 1);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(4), static_cast< GXBlendFactor >(1), static_cast< GXLogicOp >(0), 1, 0, 0);
+    GDSetGenMode2(1, 0, 1, 0, static_cast< GXCullMode >(1));
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetChanCtrl(static_cast< GXChannelID >(5), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetTevOrder(static_cast< GXTevStageID >(0), static_cast< GXTexCoordID >(0), static_cast< GXTexMapID >(0), static_cast< GXChannelID >(0xFF), static_cast< GXTexCoordID >(1), static_cast< GXTexMapID >(1), static_cast< GXChannelID >(0xFF));
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(8), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(1), static_cast< GXTevScale >(3), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 2);
 
-    GDSetAlphaCompare(7, 0, 1, 7, 0);
-    GDSetZMode(1, 6, 0);
-    GDSetGenMode2(0, 1, 1, 0, 2);
-    GDSetTevDirect(0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetBlendModeEtc(0, 0, 5, 0, 0, 1, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(1), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(6), 0);
+    GDSetGenMode2(0, 1, 1, 0, static_cast< GXCullMode >(2));
+    GDSetTevDirect(static_cast< GXTevStageID >(0));
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(0), static_cast< GXBlendFactor >(0), static_cast< GXBlendFactor >(5), static_cast< GXLogicOp >(0), 0, 1, 0);
     GDSetDstAlpha(1, 0x90);
-    GDSetTevAlphaCalcAndSwap(0, 7, 7, 7, 7, 0, 1, 3, 1, 0, 0, 0);
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(1), static_cast< GXTevScale >(3), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 3);
 
-    GDSetAlphaCompare(7, 0, 0, 7, 0);
-    GDSetZMode(1, 3, 0);
-    GDSetGenMode2(0, 1, 1, 0, 1);
-    GDSetTevDirect(0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetBlendModeEtc(1, 4, 5, 0, 1, 0, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(0), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(3), 0);
+    GDSetGenMode2(0, 1, 1, 0, static_cast< GXCullMode >(1));
+    GDSetTevDirect(static_cast< GXTevStageID >(0));
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(4), static_cast< GXBlendFactor >(5), static_cast< GXLogicOp >(0), 1, 0, 0);
     GXColor yellow = {0xFF, 0xFF, 0x00, 0xFF};
-    GDSetTevColor(1, yellow);
-    GDSetTevColorCalc(0, 2, 0xF, 0xF, 0xF, 0, 0, 0, 1, 0);
-    GDSetTevAlphaCalcAndSwap(0, 1, 7, 7, 7, 0, 0, 0, 1, 0, 0, 0);
+    GDSetTevColor(static_cast< GXTevRegID >(1), yellow);
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(2), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(1), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 4);
 
-    GDSetAlphaCompare(7, 0, 0, 7, 0);
-    GDSetZMode(1, 7, 1);
-    GDSetGenMode2(0, 1, 1, 0, 1);
-    GDSetBlendModeEtc(0, 4, 1, 0, 0, 0, 0);
-    GDSetChanCtrl(4, 0, 1, 1, 0, 2, 2);
-    GDSetTevColorCalc(0, 0xF, 0xF, 0xF, 0xF, 0, 0, 0, 1, 0);
-    GDSetTevAlphaCalcAndSwap(0, 7, 7, 7, 7, 0, 0, 0, 1, 0, 0, 0);
-    GDSetTevOrder(0, 0xFF, 0xFF, 4, 0xFF, 0xFF, 4);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(0), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(7), 1);
+    GDSetGenMode2(0, 1, 1, 0, static_cast< GXCullMode >(1));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(0), static_cast< GXBlendFactor >(4), static_cast< GXBlendFactor >(1), static_cast< GXLogicOp >(0), 0, 0, 0);
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(1), static_cast< GXColorSrc >(1), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
+    GDSetTevOrder(static_cast< GXTevStageID >(0), static_cast< GXTexCoordID >(0xFF), static_cast< GXTexMapID >(0xFF), static_cast< GXChannelID >(4), static_cast< GXTexCoordID >(0xFF), static_cast< GXTexMapID >(0xFF), static_cast< GXChannelID >(4));
     storeDisplayList(&displayList, 5);
 
-    GDSetTexLookupMode(0, 1, 1, 1, 1, 0.0f, 0.0f, 0.0f, 0, 0, 0);
-    GDSetTexImgAttr(0, static_cast< u16 >(MR::getScreenWidth()), JUTVideo::getManager()->getEfbHeight(), 4);
+    GDSetTexLookupMode(static_cast< GXTexMapID >(0), static_cast< GXTexWrapMode >(1), static_cast< GXTexWrapMode >(1), static_cast< GXTexFilter >(1), static_cast< GXTexFilter >(1), 0.0f, 0.0f, 0.0f, 0, 0, static_cast< GXAnisotropy >(0));
+    GDSetTexImgAttr(static_cast< GXTexMapID >(0), static_cast< u16 >(MR::getScreenWidth()), JUTVideo::getManager()->getEfbHeight(), static_cast< GXTexFmt >(4));
     const ResTIMG* screenImage = MR::getScreenResTIMG();
-    GDSetTexImgPtr(0, const_cast< u8* >(reinterpret_cast< const u8* >(screenImage)) + screenImage->mImageDataOffset);
-    GDSetTexCoordGen(0, 1, 1, 1, 0x3C);
+    GDSetTexImgPtr(static_cast< GXTexMapID >(0), const_cast< u8* >(reinterpret_cast< const u8* >(screenImage)) + screenImage->mImageDataOffset);
+    GDSetTexCoordGen(static_cast< GXTexCoordID >(0), static_cast< GXTexGenType >(1), static_cast< GXTexGenSrc >(1), 1, 0x3C);
     storeDisplayList(&displayList, 6);
 
     GXColor black = {0, 0, 0, 0};
     GXColor opaqueBlack = {0, 0, 0, 0xFF};
-    GDSetChanAmbColor(2, black);
-    GDSetChanMatColor(2, opaqueBlack);
-    GDSetBlendModeEtc(0, 1, 0, 5, 0, 1, 0);
-    GDSetAlphaCompare(7, 0, 0, 7, 0);
-    GDSetChanCtrl(0, 0, 0, 0, 0, 2, 2);
-    GDSetChanCtrl(2, 1, 0, 0, 1, 2, 2);
-    GDSetTevAlphaCalcAndSwap(0, 6, 7, 5, 7, 0, 0, 0, 1, 0, 0, 0);
+    GDSetChanAmbColor(static_cast< GXChannelID >(2), black);
+    GDSetChanMatColor(static_cast< GXChannelID >(2), opaqueBlack);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(0), static_cast< GXBlendFactor >(1), static_cast< GXBlendFactor >(0), static_cast< GXLogicOp >(5), 0, 1, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(0), static_cast< GXCompare >(7), 0);
+    GDSetChanCtrl(static_cast< GXChannelID >(0), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetChanCtrl(static_cast< GXChannelID >(2), 1, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 1, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(6), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(5), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 7);
 
-    GDSetCullMode(1);
-    GDSetBlendModeEtc(1, 0, 0, 5, 0, 1, 0);
-    GDSetAlphaCompare(7, 0, 1, 7, 0);
-    GDSetZMode(1, 3, 0);
+    GDSetCullMode(static_cast< GXCullMode >(1));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(0), static_cast< GXBlendFactor >(0), static_cast< GXLogicOp >(5), 0, 1, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(1), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(3), 0);
     GXColor tevColor8 = {0, 0, 0, 0xFC};
-    GDSetTevColor(1, tevColor8);
-    GDSetTevAlphaCalcAndSwap(0, 1, 7, 7, 7, 0, 0, 0, 1, 0, 0, 0);
+    GDSetTevColor(static_cast< GXTevRegID >(1), tevColor8);
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(1), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 8);
 
-    GDSetCullMode(2);
-    GDSetBlendModeEtc(1, 1, 1, 5, 0, 1, 0);
-    GDSetAlphaCompare(7, 0, 1, 7, 0);
-    GDSetZMode(1, 3, 0);
+    GDSetCullMode(static_cast< GXCullMode >(2));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(1), static_cast< GXBlendFactor >(1), static_cast< GXLogicOp >(5), 0, 1, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(1), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(3), 0);
     GXColor tevColor9 = {0, 0, 0, 4};
-    GDSetTevColor(1, tevColor9);
-    GDSetTevAlphaCalcAndSwap(0, 1, 7, 7, 7, 0, 0, 0, 1, 0, 0, 0);
+    GDSetTevColor(static_cast< GXTevRegID >(1), tevColor9);
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(1), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 9);
 
-    GDSetBlendModeEtc(1, 6, 7, 5, 1, 1, 0);
-    GDSetZMode(1, 3, 0);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(6), static_cast< GXBlendFactor >(7), static_cast< GXLogicOp >(5), 1, 1, 0);
+    GDSetZMode(1, static_cast< GXCompare >(3), 0);
     GDSetDstAlpha(1, 0);
     storeDisplayList(&displayList, 10);
 
-    GDSetZMode(1, 3, 1);
-    GDSetGenMode2(1, 0, 1, 0, 1);
-    GDSetAlphaCompare(4, 0x20, 0, 7, 0);
-    GDSetBlendModeEtc(0, 0, 0, 5, 1, 1, 0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetChanCtrl(5, 0, 0, 0, 0, 2, 2);
-    GDSetTevColorCalc(0, 8, 0xF, 0xF, 0xF, 0, 0, 3, 1, 0);
-    GDSetTevOrder(0, 0, 0, 0xFF, 1, 1, 0xFF);
+    GDSetZMode(1, static_cast< GXCompare >(3), 1);
+    GDSetGenMode2(1, 0, 1, 0, static_cast< GXCullMode >(1));
+    GDSetAlphaCompare(static_cast< GXCompare >(4), 0x20, static_cast< GXAlphaOp >(0), static_cast< GXCompare >(7), 0);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(0), static_cast< GXBlendFactor >(0), static_cast< GXBlendFactor >(0), static_cast< GXLogicOp >(5), 1, 1, 0);
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetChanCtrl(static_cast< GXChannelID >(5), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(8), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(3), 1, static_cast< GXTevRegID >(0));
+    GDSetTevOrder(static_cast< GXTevStageID >(0), static_cast< GXTexCoordID >(0), static_cast< GXTexMapID >(0), static_cast< GXChannelID >(0xFF), static_cast< GXTexCoordID >(1), static_cast< GXTexMapID >(1), static_cast< GXChannelID >(0xFF));
     storeDisplayList(&displayList, 11);
 
-    GDSetZMode(1, 3, 0);
-    GDSetGenMode2(1, 0, 1, 0, 2);
-    GDSetAlphaCompare(7, 0, 0, 7, 0);
-    GDSetBlendModeEtc(1, 4, 1, 5, 1, 0, 0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetChanCtrl(5, 0, 0, 0, 0, 2, 2);
-    GDSetTevColorCalc(0, 0xF, 0xC, 2, 0xF, 0, 0, 0, 1, 0);
-    GDSetTevAlphaCalcAndSwap(0, 1, 7, 7, 7, 0, 0, 0, 1, 0, 0, 0);
-    GDSetTevOrder(0, 0, 0, 0xFF, 1, 1, 0xFF);
+    GDSetZMode(1, static_cast< GXCompare >(3), 0);
+    GDSetGenMode2(1, 0, 1, 0, static_cast< GXCullMode >(2));
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(0), static_cast< GXCompare >(7), 0);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(4), static_cast< GXBlendFactor >(1), static_cast< GXLogicOp >(5), 1, 0, 0);
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetChanCtrl(static_cast< GXChannelID >(5), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xC), static_cast< GXTevColorArg >(2), static_cast< GXTevColorArg >(0xF), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(1), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
+    GDSetTevOrder(static_cast< GXTevStageID >(0), static_cast< GXTexCoordID >(0), static_cast< GXTexMapID >(0), static_cast< GXChannelID >(0xFF), static_cast< GXTexCoordID >(1), static_cast< GXTexMapID >(1), static_cast< GXChannelID >(0xFF));
     storeDisplayList(&displayList, 12);
 
-    GDSetAlphaCompare(7, 0, 1, 7, 0);
-    GDSetZMode(1, 6, 0);
-    GDSetGenMode2(0, 1, 1, 0, 2);
-    GDSetTevDirect(0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetBlendModeEtc(1, 1, 4, 0, 1, 0, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(1), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(6), 0);
+    GDSetGenMode2(0, 1, 1, 0, static_cast< GXCullMode >(2));
+    GDSetTevDirect(static_cast< GXTevStageID >(0));
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(1), static_cast< GXBlendFactor >(4), static_cast< GXLogicOp >(0), 1, 0, 0);
     GXColor tevColor13 = {0x23, 0x19, 0x19, 5};
-    GDSetTevColor(1, tevColor13);
-    GDSetTevColorCalc(0, 0xF, 0xF, 0xF, 2, 0, 0, 0, 1, 0);
-    GDSetTevAlphaCalcAndSwap(0, 7, 7, 7, 1, 0, 0, 0, 1, 0, 0, 0);
+    GDSetTevColor(static_cast< GXTevRegID >(1), tevColor13);
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(2), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(1), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 13);
 
     GDSetDstAlpha(1, 0x40);
     storeDisplayList(&displayList, 14);
 
-    GDSetBlendModeEtc(1, 7, 6, 0, 1, 0, 0);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(7), static_cast< GXBlendFactor >(6), static_cast< GXLogicOp >(0), 1, 0, 0);
     storeDisplayList(&displayList, 15);
 
     _1E0 = -1;
@@ -495,7 +481,7 @@
     _1BC = 0;
 
     const u16 materialCount = mModelData->getMaterialNum();
-    _1C4 = new u32[materialCount];
+    _1C4 = new u8*[materialCount];
     _1C8 = new u8*[materialCount];
     _1CC = new u16[materialCount];
 
@@ -568,5 +554,7 @@
 J3DModelX::~J3DModelX() {
 }
 
+#ifndef TARGET_PC
 J3DModel::~J3DModel() {
 }
+#endif
--- a/pc-port/src/Game/Player/J3DModelX.cpp
+++ b/pc-port/src/Game/Player/J3DModelX.cpp
@@ -18,26 +18,12 @@
 #include <revolution/gd/GDBase.h>
 #include <revolution/gx.h>
 
-extern "C" {
-void GDSetCullMode(int);
-void GDSetAlphaCompare(int, int, int, int, int);
-void GDSetZMode(int, int, int);
-void GDSetGenMode2(int, int, int, int, int);
-void GDSetTevDirect(int);
-void GDSetChanCtrl(int, int, int, int, int, int, int);
-void GDSetBlendModeEtc(int, int, int, int, int, int, int);
-void GDSetDstAlpha(int, int);
-void GDSetTevAlphaCalcAndSwap(int, int, int, int, int, int, int, int, int, int, int, int);
-void GDSetTevOrder(int, int, int, int, int, int, int);
-void GDSetTevColorCalc(int, int, int, int, int, int, int, int, int, int);
-void GDSetTevColor(int, GXColor);
-void GDSetTexLookupMode(int, int, int, int, int, f32, f32, f32, int, int, int);
-void GDSetTexImgAttr(int, int, int, int);
-void GDSetTexImgPtr(int, void*);
-void GDSetTexCoordGen(int, int, int, int, int);
-void GDSetChanAmbColor(int, GXColor);
-void GDSetChanMatColor(int, GXColor);
-}
+#include <revolution/gd/GDGeometry.h>
+#include <revolution/gd/GDTev.h>
+#include <revolution/gd/GDPixel.h>
+#include <revolution/gd/GDIndirect.h>
+#include <revolution/gd/GDLight.h>
+#include <revolution/gd/GDTexture.h>
 
 class J3DMtxBuffer2 : public J3DMtxBuffer {
 public:
@@ -259,30 +245,30 @@
     }
 
     if (_1B8 != nullptr) {
-        _1B4 = reinterpret_cast< u32 >(_1B8);
+        _1B4 = _1B8;
         _1B8 = nullptr;
     }
 
     if (_1B4 != 0 && _1BC != 0) {
-        GXCallDisplayList(reinterpret_cast< void* >(_1B4), _1BC);
+        GXCallDisplayList(_1B4, _1BC);
     }
 
     if (_1C8[_1C0] != nullptr) {
-        _1C4[_1C0] = reinterpret_cast< u32 >(_1C8[_1C0]);
+        _1C4[_1C0] = _1C8[_1C0];
         _1C8[_1C0] = nullptr;
     }
 
     if (_1C4[_1C0] != 0 && _1CC[_1C0] != 0) {
-        GXCallDisplayList(reinterpret_cast< void* >(_1C4[_1C0]), _1CC[_1C0]);
+        GXCallDisplayList(_1C4[_1C0], _1CC[_1C0]);
     }
 
     if (mMaterialCallback != nullptr) {
-        mMaterialCallback(reinterpret_cast< void* >(_128), _1C0);
+        mMaterialCallback(_128, _1C0);
     }
 
     J3DShape* originalShape = nullptr;
     if (_12C != 0) {
-        J3DModel* shapeModel = reinterpret_cast< J3DModel* >(_12C);
+        J3DModel* shapeModel = _12C;
         originalShape = shapePacket->mpShape;
         J3DShape* replacementShape = shapeModel->mShapePacket[material->mShape->mIndex].mpShape;
         shapePacket->mpShape = replacementShape;
@@ -348,140 +334,140 @@
     GDInitGDLObj(&displayList, displayListBuffer, sizeof(displayListBuffer));
     __GDCurrentDL = &displayList;
 
-    GDSetCullMode(1);
+    GDSetCullMode(static_cast< GXCullMode >(1));
     storeDisplayList(&displayList, 0);
 
-    GDSetAlphaCompare(7, 0, 1, 7, 0);
-    GDSetZMode(0, 7, 0);
-    GDSetGenMode2(0, 1, 1, 0, 2);
-    GDSetTevDirect(0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetBlendModeEtc(0, 0, 5, 0, 0, 1, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(1), static_cast< GXCompare >(7), 0);
+    GDSetZMode(0, static_cast< GXCompare >(7), 0);
+    GDSetGenMode2(0, 1, 1, 0, static_cast< GXCullMode >(2));
+    GDSetTevDirect(static_cast< GXTevStageID >(0));
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(0), static_cast< GXBlendFactor >(0), static_cast< GXBlendFactor >(5), static_cast< GXLogicOp >(0), 0, 1, 0);
     GDSetDstAlpha(1, 0xFF);
-    GDSetTevAlphaCalcAndSwap(0, 7, 7, 7, 7, 0, 1, 3, 1, 0, 0, 0);
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(1), static_cast< GXTevScale >(3), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 1);
 
-    GDSetCullMode(1);
-    GDSetZMode(1, 3, 1);
-    GDSetBlendModeEtc(1, 4, 1, 0, 1, 0, 0);
-    GDSetGenMode2(1, 0, 1, 0, 1);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetChanCtrl(5, 0, 0, 0, 0, 2, 2);
-    GDSetTevOrder(0, 0, 0, 0xFF, 1, 1, 0xFF);
-    GDSetTevColorCalc(0, 8, 0xF, 0xF, 0xF, 0, 0, 0, 1, 0);
-    GDSetTevAlphaCalcAndSwap(0, 7, 7, 7, 7, 0, 1, 3, 1, 0, 0, 0);
+    GDSetCullMode(static_cast< GXCullMode >(1));
+    GDSetZMode(1, static_cast< GXCompare >(3), 1);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(4), static_cast< GXBlendFactor >(1), static_cast< GXLogicOp >(0), 1, 0, 0);
+    GDSetGenMode2(1, 0, 1, 0, static_cast< GXCullMode >(1));
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetChanCtrl(static_cast< GXChannelID >(5), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetTevOrder(static_cast< GXTevStageID >(0), static_cast< GXTexCoordID >(0), static_cast< GXTexMapID >(0), static_cast< GXChannelID >(0xFF), static_cast< GXTexCoordID >(1), static_cast< GXTexMapID >(1), static_cast< GXChannelID >(0xFF));
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(8), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(1), static_cast< GXTevScale >(3), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 2);
 
-    GDSetAlphaCompare(7, 0, 1, 7, 0);
-    GDSetZMode(1, 6, 0);
-    GDSetGenMode2(0, 1, 1, 0, 2);
-    GDSetTevDirect(0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetBlendModeEtc(0, 0, 5, 0, 0, 1, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(1), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(6), 0);
+    GDSetGenMode2(0, 1, 1, 0, static_cast< GXCullMode >(2));
+    GDSetTevDirect(static_cast< GXTevStageID >(0));
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(0), static_cast< GXBlendFactor >(0), static_cast< GXBlendFactor >(5), static_cast< GXLogicOp >(0), 0, 1, 0);
     GDSetDstAlpha(1, 0x90);
-    GDSetTevAlphaCalcAndSwap(0, 7, 7, 7, 7, 0, 1, 3, 1, 0, 0, 0);
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(1), static_cast< GXTevScale >(3), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 3);
 
-    GDSetAlphaCompare(7, 0, 0, 7, 0);
-    GDSetZMode(1, 3, 0);
-    GDSetGenMode2(0, 1, 1, 0, 1);
-    GDSetTevDirect(0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetBlendModeEtc(1, 4, 5, 0, 1, 0, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(0), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(3), 0);
+    GDSetGenMode2(0, 1, 1, 0, static_cast< GXCullMode >(1));
+    GDSetTevDirect(static_cast< GXTevStageID >(0));
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(4), static_cast< GXBlendFactor >(5), static_cast< GXLogicOp >(0), 1, 0, 0);
     GXColor yellow = {0xFF, 0xFF, 0x00, 0xFF};
-    GDSetTevColor(1, yellow);
-    GDSetTevColorCalc(0, 2, 0xF, 0xF, 0xF, 0, 0, 0, 1, 0);
-    GDSetTevAlphaCalcAndSwap(0, 1, 7, 7, 7, 0, 0, 0, 1, 0, 0, 0);
+    GDSetTevColor(static_cast< GXTevRegID >(1), yellow);
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(2), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(1), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 4);
 
-    GDSetAlphaCompare(7, 0, 0, 7, 0);
-    GDSetZMode(1, 7, 1);
-    GDSetGenMode2(0, 1, 1, 0, 1);
-    GDSetBlendModeEtc(0, 4, 1, 0, 0, 0, 0);
-    GDSetChanCtrl(4, 0, 1, 1, 0, 2, 2);
-    GDSetTevColorCalc(0, 0xF, 0xF, 0xF, 0xF, 0, 0, 0, 1, 0);
-    GDSetTevAlphaCalcAndSwap(0, 7, 7, 7, 7, 0, 0, 0, 1, 0, 0, 0);
-    GDSetTevOrder(0, 0xFF, 0xFF, 4, 0xFF, 0xFF, 4);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(0), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(7), 1);
+    GDSetGenMode2(0, 1, 1, 0, static_cast< GXCullMode >(1));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(0), static_cast< GXBlendFactor >(4), static_cast< GXBlendFactor >(1), static_cast< GXLogicOp >(0), 0, 0, 0);
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(1), static_cast< GXColorSrc >(1), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
+    GDSetTevOrder(static_cast< GXTevStageID >(0), static_cast< GXTexCoordID >(0xFF), static_cast< GXTexMapID >(0xFF), static_cast< GXChannelID >(4), static_cast< GXTexCoordID >(0xFF), static_cast< GXTexMapID >(0xFF), static_cast< GXChannelID >(4));
     storeDisplayList(&displayList, 5);
 
-    GDSetTexLookupMode(0, 1, 1, 1, 1, 0.0f, 0.0f, 0.0f, 0, 0, 0);
-    GDSetTexImgAttr(0, static_cast< u16 >(MR::getScreenWidth()), JUTVideo::getManager()->getEfbHeight(), 4);
+    GDSetTexLookupMode(static_cast< GXTexMapID >(0), static_cast< GXTexWrapMode >(1), static_cast< GXTexWrapMode >(1), static_cast< GXTexFilter >(1), static_cast< GXTexFilter >(1), 0.0f, 0.0f, 0.0f, 0, 0, static_cast< GXAnisotropy >(0));
+    GDSetTexImgAttr(static_cast< GXTexMapID >(0), static_cast< u16 >(MR::getScreenWidth()), JUTVideo::getManager()->getEfbHeight(), static_cast< GXTexFmt >(4));
     const ResTIMG* screenImage = MR::getScreenResTIMG();
-    GDSetTexImgPtr(0, const_cast< u8* >(reinterpret_cast< const u8* >(screenImage)) + screenImage->mImageDataOffset);
-    GDSetTexCoordGen(0, 1, 1, 1, 0x3C);
+    GDSetTexImgPtr(static_cast< GXTexMapID >(0), const_cast< u8* >(reinterpret_cast< const u8* >(screenImage)) + screenImage->mImageDataOffset);
+    GDSetTexCoordGen(static_cast< GXTexCoordID >(0), static_cast< GXTexGenType >(1), static_cast< GXTexGenSrc >(1), 1, 0x3C);
     storeDisplayList(&displayList, 6);
 
     GXColor black = {0, 0, 0, 0};
     GXColor opaqueBlack = {0, 0, 0, 0xFF};
-    GDSetChanAmbColor(2, black);
-    GDSetChanMatColor(2, opaqueBlack);
-    GDSetBlendModeEtc(0, 1, 0, 5, 0, 1, 0);
-    GDSetAlphaCompare(7, 0, 0, 7, 0);
-    GDSetChanCtrl(0, 0, 0, 0, 0, 2, 2);
-    GDSetChanCtrl(2, 1, 0, 0, 1, 2, 2);
-    GDSetTevAlphaCalcAndSwap(0, 6, 7, 5, 7, 0, 0, 0, 1, 0, 0, 0);
+    GDSetChanAmbColor(static_cast< GXChannelID >(2), black);
+    GDSetChanMatColor(static_cast< GXChannelID >(2), opaqueBlack);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(0), static_cast< GXBlendFactor >(1), static_cast< GXBlendFactor >(0), static_cast< GXLogicOp >(5), 0, 1, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(0), static_cast< GXCompare >(7), 0);
+    GDSetChanCtrl(static_cast< GXChannelID >(0), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetChanCtrl(static_cast< GXChannelID >(2), 1, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 1, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(6), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(5), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 7);
 
-    GDSetCullMode(1);
-    GDSetBlendModeEtc(1, 0, 0, 5, 0, 1, 0);
-    GDSetAlphaCompare(7, 0, 1, 7, 0);
-    GDSetZMode(1, 3, 0);
+    GDSetCullMode(static_cast< GXCullMode >(1));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(0), static_cast< GXBlendFactor >(0), static_cast< GXLogicOp >(5), 0, 1, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(1), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(3), 0);
     GXColor tevColor8 = {0, 0, 0, 0xFC};
-    GDSetTevColor(1, tevColor8);
-    GDSetTevAlphaCalcAndSwap(0, 1, 7, 7, 7, 0, 0, 0, 1, 0, 0, 0);
+    GDSetTevColor(static_cast< GXTevRegID >(1), tevColor8);
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(1), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 8);
 
-    GDSetCullMode(2);
-    GDSetBlendModeEtc(1, 1, 1, 5, 0, 1, 0);
-    GDSetAlphaCompare(7, 0, 1, 7, 0);
-    GDSetZMode(1, 3, 0);
+    GDSetCullMode(static_cast< GXCullMode >(2));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(1), static_cast< GXBlendFactor >(1), static_cast< GXLogicOp >(5), 0, 1, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(1), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(3), 0);
     GXColor tevColor9 = {0, 0, 0, 4};
-    GDSetTevColor(1, tevColor9);
-    GDSetTevAlphaCalcAndSwap(0, 1, 7, 7, 7, 0, 0, 0, 1, 0, 0, 0);
+    GDSetTevColor(static_cast< GXTevRegID >(1), tevColor9);
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(1), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 9);
 
-    GDSetBlendModeEtc(1, 6, 7, 5, 1, 1, 0);
-    GDSetZMode(1, 3, 0);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(6), static_cast< GXBlendFactor >(7), static_cast< GXLogicOp >(5), 1, 1, 0);
+    GDSetZMode(1, static_cast< GXCompare >(3), 0);
     GDSetDstAlpha(1, 0);
     storeDisplayList(&displayList, 10);
 
-    GDSetZMode(1, 3, 1);
-    GDSetGenMode2(1, 0, 1, 0, 1);
-    GDSetAlphaCompare(4, 0x20, 0, 7, 0);
-    GDSetBlendModeEtc(0, 0, 0, 5, 1, 1, 0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetChanCtrl(5, 0, 0, 0, 0, 2, 2);
-    GDSetTevColorCalc(0, 8, 0xF, 0xF, 0xF, 0, 0, 3, 1, 0);
-    GDSetTevOrder(0, 0, 0, 0xFF, 1, 1, 0xFF);
+    GDSetZMode(1, static_cast< GXCompare >(3), 1);
+    GDSetGenMode2(1, 0, 1, 0, static_cast< GXCullMode >(1));
+    GDSetAlphaCompare(static_cast< GXCompare >(4), 0x20, static_cast< GXAlphaOp >(0), static_cast< GXCompare >(7), 0);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(0), static_cast< GXBlendFactor >(0), static_cast< GXBlendFactor >(0), static_cast< GXLogicOp >(5), 1, 1, 0);
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetChanCtrl(static_cast< GXChannelID >(5), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(8), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(3), 1, static_cast< GXTevRegID >(0));
+    GDSetTevOrder(static_cast< GXTevStageID >(0), static_cast< GXTexCoordID >(0), static_cast< GXTexMapID >(0), static_cast< GXChannelID >(0xFF), static_cast< GXTexCoordID >(1), static_cast< GXTexMapID >(1), static_cast< GXChannelID >(0xFF));
     storeDisplayList(&displayList, 11);
 
-    GDSetZMode(1, 3, 0);
-    GDSetGenMode2(1, 0, 1, 0, 2);
-    GDSetAlphaCompare(7, 0, 0, 7, 0);
-    GDSetBlendModeEtc(1, 4, 1, 5, 1, 0, 0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetChanCtrl(5, 0, 0, 0, 0, 2, 2);
-    GDSetTevColorCalc(0, 0xF, 0xC, 2, 0xF, 0, 0, 0, 1, 0);
-    GDSetTevAlphaCalcAndSwap(0, 1, 7, 7, 7, 0, 0, 0, 1, 0, 0, 0);
-    GDSetTevOrder(0, 0, 0, 0xFF, 1, 1, 0xFF);
+    GDSetZMode(1, static_cast< GXCompare >(3), 0);
+    GDSetGenMode2(1, 0, 1, 0, static_cast< GXCullMode >(2));
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(0), static_cast< GXCompare >(7), 0);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(4), static_cast< GXBlendFactor >(1), static_cast< GXLogicOp >(5), 1, 0, 0);
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetChanCtrl(static_cast< GXChannelID >(5), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xC), static_cast< GXTevColorArg >(2), static_cast< GXTevColorArg >(0xF), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(1), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
+    GDSetTevOrder(static_cast< GXTevStageID >(0), static_cast< GXTexCoordID >(0), static_cast< GXTexMapID >(0), static_cast< GXChannelID >(0xFF), static_cast< GXTexCoordID >(1), static_cast< GXTexMapID >(1), static_cast< GXChannelID >(0xFF));
     storeDisplayList(&displayList, 12);
 
-    GDSetAlphaCompare(7, 0, 1, 7, 0);
-    GDSetZMode(1, 6, 0);
-    GDSetGenMode2(0, 1, 1, 0, 2);
-    GDSetTevDirect(0);
-    GDSetChanCtrl(4, 0, 0, 0, 0, 2, 2);
-    GDSetBlendModeEtc(1, 1, 4, 0, 1, 0, 0);
+    GDSetAlphaCompare(static_cast< GXCompare >(7), 0, static_cast< GXAlphaOp >(1), static_cast< GXCompare >(7), 0);
+    GDSetZMode(1, static_cast< GXCompare >(6), 0);
+    GDSetGenMode2(0, 1, 1, 0, static_cast< GXCullMode >(2));
+    GDSetTevDirect(static_cast< GXTevStageID >(0));
+    GDSetChanCtrl(static_cast< GXChannelID >(4), 0, static_cast< GXColorSrc >(0), static_cast< GXColorSrc >(0), 0, static_cast< GXDiffuseFn >(2), static_cast< GXAttnFn >(2));
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(1), static_cast< GXBlendFactor >(4), static_cast< GXLogicOp >(0), 1, 0, 0);
     GXColor tevColor13 = {0x23, 0x19, 0x19, 5};
-    GDSetTevColor(1, tevColor13);
-    GDSetTevColorCalc(0, 0xF, 0xF, 0xF, 2, 0, 0, 0, 1, 0);
-    GDSetTevAlphaCalcAndSwap(0, 7, 7, 7, 1, 0, 0, 0, 1, 0, 0, 0);
+    GDSetTevColor(static_cast< GXTevRegID >(1), tevColor13);
+    GDSetTevColorCalc(static_cast< GXTevStageID >(0), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(0xF), static_cast< GXTevColorArg >(2), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0));
+    GDSetTevAlphaCalcAndSwap(static_cast< GXTevStageID >(0), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(7), static_cast< GXTevAlphaArg >(1), static_cast< GXTevOp >(0), static_cast< GXTevBias >(0), static_cast< GXTevScale >(0), 1, static_cast< GXTevRegID >(0), static_cast< GXTevSwapSel >(0), static_cast< GXTevSwapSel >(0));
     storeDisplayList(&displayList, 13);
 
     GDSetDstAlpha(1, 0x40);
     storeDisplayList(&displayList, 14);
 
-    GDSetBlendModeEtc(1, 7, 6, 0, 1, 0, 0);
+    GDSetBlendModeEtc(static_cast< GXBlendMode >(1), static_cast< GXBlendFactor >(7), static_cast< GXBlendFactor >(6), static_cast< GXLogicOp >(0), 1, 0, 0);
     storeDisplayList(&displayList, 15);
 
     _1E0 = -1;
@@ -495,7 +481,7 @@
     _1BC = 0;
 
     const u16 materialCount = mModelData->getMaterialNum();
-    _1C4 = new u32[materialCount];
+    _1C4 = new u8*[materialCount];
     _1C8 = new u8*[materialCount];
     _1CC = new u16[materialCount];
 
@@ -568,5 +554,7 @@
 J3DModelX::~J3DModelX() {
 }
 
+#ifndef TARGET_PC
 J3DModel::~J3DModel() {
 }
+#endif
--- a/include/Game/Player/J3DModelX.hpp
+++ b/include/Game/Player/J3DModelX.hpp
@@ -37,6 +37,38 @@
             *(u32*)this = 0;
         }
 
+#if defined(TARGET_PC) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+        unsigned : 3;
+        unsigned _1C : 1;
+        unsigned _1B : 1;
+        unsigned _1A : 1;
+        unsigned _19 : 1;
+        unsigned _18 : 1;
+        unsigned _17 : 1;
+        unsigned _16 : 1;
+        unsigned _15 : 1;
+        unsigned _14 : 1;
+        unsigned _13 : 1;
+        unsigned _12 : 1;
+        unsigned _11 : 1;
+        unsigned _10 : 1;
+        unsigned _F : 1;
+        unsigned _E : 1;
+        unsigned _D : 1;
+        unsigned _C : 1;
+        unsigned _B : 1;
+        unsigned _A : 1;
+        unsigned _9 : 1;
+        unsigned _8 : 1;
+        unsigned _7 : 1;
+        unsigned _6 : 1;
+        unsigned _5 : 1;
+        unsigned _4 : 1;
+        unsigned _3 : 1;
+        unsigned _2 : 1;
+        unsigned _1 : 1;
+        unsigned _0 : 1;
+#else
         unsigned _0 : 1;
         unsigned _1 : 1;
         unsigned _2 : 1;
@@ -66,6 +98,7 @@
         unsigned _1A : 1;
         unsigned _1B : 1;
         unsigned _1C : 1;
+#endif
     };
 
     u8 _DC;
@@ -93,17 +126,17 @@
     };
     void (*mMaterialCallback)(void*, u16);
     void (*mShapeCallback)(J3DShape*);
-    u32 _128;
-    u32 _12C;
+    void* _128;
+    J3DModel* _12C;
     u8* mDisplayLists[16];
     u32 mDisplayListSizes[16];
     Flags mFlags;
-    u32 _1B4;
+    u8* _1B4;
     u8* _1B8;
     u32 _1BC;
     u16 _1C0;
     u16 _1C2;
-    u32* _1C4;
+    u8** _1C4;
     u8** _1C8;
     u16* _1CC;
     u8 _1D0;
--- /dev/null
+++ b/pc-port/src/compat/J3DModelXDrawCompat.cpp
@@ -0,0 +1,27 @@
+#include "Game/Player/J3DModelX.hpp"
+
+void J3DModelX::copyExtraMtxBuffer(const J3DModelX* pModel) {
+    _DD = pModel->_DD;
+
+    for (u32 i = 0; i < _DD; i++) {
+        mExtraMtxBuffer[i] = pModel->mExtraMtxBuffer[i];
+    }
+}
+
+void J3DModelX::swapDrawBuffer(u32 drawBuffer) {
+    if (_DC == drawBuffer) {
+        _DC = 0;
+        return;
+    }
+
+    _DC = drawBuffer;
+}
+
+void J3DModelX::setDynamicDL(u8* pDL, u32 dlSize) {
+    if (pDL == nullptr) {
+        dlSize = 0;
+    }
+
+    _1B8 = pDL;
+    _1BC = dlSize;
+}
--- /dev/null
+++ b/pc-port/src/compat/ModelCreationCompat.cpp
@@ -0,0 +1,89 @@
+#include "Game/Util/ModelUtil.hpp"
+#include "Game/Animation/XanimePlayer.hpp"
+#include "Game/Animation/XanimeResource.hpp"
+#include "Game/Player/J3DModelX.hpp"
+#include "Game/System/ResourceHolder.hpp"
+#include "Game/System/ShapePacketUserData.hpp"
+#include "Game/Util/MutexHolder.hpp"
+#include "Game/Util/SchedulerUtil.hpp"
+#include "Game/Util/StringUtil.hpp"
+#include <JSystem/J3DGraphAnimator/J3DJoint.hpp>
+#include <JSystem/J3DGraphBase/J3DShape.hpp>
+#include <cstring>
+
+namespace MR {
+
+void invalidateMtxCalc(J3DModelData* pModelData) {
+        u16 jointNum = pModelData->getJointNum();
+        for (u16 i = 0; i < jointNum; i++) {
+            pModelData->getJointNodePointer(i)->setMtxCalc(nullptr);
+        }
+    }
+
+void invalidateJointCallback(J3DModelData* pModelData) {
+        u16 jointNum = pModelData->getJointNum();
+        for (u16 i = 0; i < jointNum; i++) {
+            pModelData->getJointNodePointer(i)->setCallBack(nullptr);
+        }
+    }
+
+J3DModel* newJ3DModel(const ResourceHolder* pResourceHolder, const char* pChar, J3DMdlFlag mdlFlag) {
+        J3DModelData* modelData = static_cast< J3DModelData* >(pResourceHolder->mModelResTable->getRes(pChar));
+        J3DModel* newModel;
+        invalidateMtxCalc(modelData);
+        invalidateJointCallback(modelData);
+
+        for (u16 i = 0; i < modelData->getShapeNum(); i++) {
+            modelData->getShapeNodePointer(i)->mFlags &= ~0x1;
+        }
+
+        if (MR::isEqualString(pChar, "Mario") || MR::isEqualString(pChar, "Luigi") || MR::isEqualString(pChar, "MarioHandL") ||
+            MR::isEqualString(pChar, "MarioHandR") || MR::isEqualString(pChar, "MarioFace") || MR::isEqualString(pChar, "LuigiFace") ||
+            MR::isEqualString(pChar, "MarioShadow") || MR::isEqualString(pChar, "LuigiShadow") || MR::isEqualString(pChar, "IceMario") ||
+            MR::isEqualString(pChar, "IceLuigi") || MR::isEqualString(pChar, "IceMarioHandL") || MR::isEqualString(pChar, "IceMarioHandR") ||
+            MR::isEqualString(pChar, "InvincibleMarioHandL") || MR::isEqualString(pChar, "InvincibleMarioHandR") ||
+            MR::isEqualString(pChar, "InvincibleMario") || MR::isEqualString(pChar, "InvincibleLuigi") || MR::isEqualString(pChar, "BeeMario") ||
+            MR::isEqualString(pChar, "BeeLuigi") || MR::isEqualString(pChar, "HopperMario") || MR::isEqualString(pChar, "HopperLuigi") ||
+            MR::isEqualString(pChar, "BoneMario") || MR::isEqualString(pChar, "BoneLuigi") || MR::isEqualString(pChar, "SearchLightCone") ||
+            strstr(pChar, "FlexibleSphere") || strstr(pChar, "FlexibleSandPlanetParts") || MR::isEqualString(pChar, "ScaleDownRelayPlanet") ||
+            MR::isEqualString(pChar, "GhostMario") || MR::isEqualString(pChar, "GhostLuigi") || MR::isEqualString(pChar, "Koura")) {
+            OSLockMutex(&MR::MutexHolder< 0 >::sMutex);
+            newModel = static_cast< J3DModel* >(new J3DModelX(modelData, mdlFlag, 1));
+            OSUnlockMutex(&MR::MutexHolder< 0 >::sMutex);
+        } else {
+            OSLockMutex(&MR::MutexHolder< 0 >::sMutex);
+            newModel = new J3DModel(modelData, mdlFlag, 1);
+            OSUnlockMutex(&MR::MutexHolder< 0 >::sMutex);
+        }
+        initJ3DShapePacketUserData(newModel);
+        Mtx mtx;
+        PSMTXIdentity(mtx);
+        newModel->unlock();
+        OSLockMutex(&MR::MutexHolder< 0 >::sMutex);
+        PSMTXCopy(mtx, j3dSys.mViewMtx);
+        newModel->calc();
+        newModel->calcMaterial();
+        newModel->viewCalc();
+        newModel->viewCalc();
+        OSUnlockMutex(&MR::MutexHolder< 0 >::sMutex);
+        if (modelData->getModelDataType() == 0) {
+            ProhibitSchedulerAndInterrupts prohibit(false);
+            OSLockMutex(&MR::MutexHolder< 0 >::sMutex);
+            newModel->makeDL();
+            OSUnlockMutex(&MR::MutexHolder< 0 >::sMutex);
+        }
+        newModel->lock();
+        return newModel;
+    }
+
+XanimePlayer* newXanimePlayer(const ResourceHolder* pModelResourceHolder, const char* pChar, const ResourceHolder* pUnused, J3DMdlFlag mdlFlag,
+                                  XanimeResourceTable* pAnimeResourceTable) {
+        J3DModel* newModel = newJ3DModel(pModelResourceHolder, pChar, mdlFlag);
+        return new XanimePlayer(newModel, pAnimeResourceTable);
+    }
+
+XanimeResourceTable* newXanimeResourceTable(ResourceHolder* pResourceHolder) {
+        return new XanimeResourceTable(pResourceHolder);
+    }
+
+}
--- /dev/null
+++ b/pc-port/src/compat/ModelFogCompat.cpp
@@ -0,0 +1,50 @@
+#include "Game/Util/DirectDraw.hpp"
+#include "Game/Util/CameraUtil.hpp"
+#include "Game/Util/ModelUtil.hpp"
+#include <JSystem/J3DGraphBase/J3DSys.hpp>
+
+namespace TDDraw {
+
+void mixFogColor(TVec3f a1, f32 a2, u32 a3) {
+        f32 nearZ = MR::getNearZ();
+        f32 farZ = MR::getFarZ();
+        GXColor color;
+        setGXColor(a3, &color);
+        f32 v11;
+        f32 v10;
+        MR::calcFogStartEnd(a1, a2, &v11, &v10);
+        GXSetFog(GX_FOG_PERSP_LIN, v11, v10, nearZ, farZ, color);
+    }
+
+void setGXColor(u32 a1, GXColor* pColor) {
+        pColor->r = (a1 >> 24) & 0xFF;
+        pColor->g = (a1 >> 16) & 0xFF;
+        pColor->b = (a1 >> 8) & 0xFF;
+        pColor->a = a1 & 0xFF;
+    }
+
+}
+
+namespace MR {
+void calcFogStartEnd(TVec3f vec, f32 f1, f32* pFogStart, f32* pFogEnd) {
+        Vec multVec;
+        PSMTXMultVec(j3dSys.mViewMtx, vec, &multVec);
+
+        f32 flt = -multVec.z;
+        if (f1 == 0.0f) {
+            *pFogStart = flt;
+            *pFogEnd = 100000.0f;
+            return;
+        }
+
+        f32 fogStart = 100.0f;
+        f32 fogEnd = 100.0f + ((flt - 100.0f) / f1);
+        if (fogEnd > 100000.0f) {
+            fogEnd = 100000.0f;
+            fogStart = flt - (100000.0f - flt) * (f1 / (1.0f - f1));
+        }
+
+        *pFogStart = fogStart;
+        *pFogEnd = fogEnd;
+    }
+}
--- a/pc-port/src/Game/Player/J3DModelX.hpp
+++ b/pc-port/src/Game/Player/J3DModelX.hpp
@@ -37,6 +37,38 @@
             *(u32*)this = 0;
         }
 
+#if defined(TARGET_PC) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+        unsigned : 3;
+        unsigned _1C : 1;
+        unsigned _1B : 1;
+        unsigned _1A : 1;
+        unsigned _19 : 1;
+        unsigned _18 : 1;
+        unsigned _17 : 1;
+        unsigned _16 : 1;
+        unsigned _15 : 1;
+        unsigned _14 : 1;
+        unsigned _13 : 1;
+        unsigned _12 : 1;
+        unsigned _11 : 1;
+        unsigned _10 : 1;
+        unsigned _F : 1;
+        unsigned _E : 1;
+        unsigned _D : 1;
+        unsigned _C : 1;
+        unsigned _B : 1;
+        unsigned _A : 1;
+        unsigned _9 : 1;
+        unsigned _8 : 1;
+        unsigned _7 : 1;
+        unsigned _6 : 1;
+        unsigned _5 : 1;
+        unsigned _4 : 1;
+        unsigned _3 : 1;
+        unsigned _2 : 1;
+        unsigned _1 : 1;
+        unsigned _0 : 1;
+#else
         unsigned _0 : 1;
         unsigned _1 : 1;
         unsigned _2 : 1;
@@ -66,6 +98,7 @@
         unsigned _1A : 1;
         unsigned _1B : 1;
         unsigned _1C : 1;
+#endif
     };
 
     u8 _DC;
@@ -93,17 +126,17 @@
     };
     void (*mMaterialCallback)(void*, u16);
     void (*mShapeCallback)(J3DShape*);
-    u32 _128;
-    u32 _12C;
+    void* _128;
+    J3DModel* _12C;
     u8* mDisplayLists[16];
     u32 mDisplayListSizes[16];
     Flags mFlags;
-    u32 _1B4;
+    u8* _1B4;
     u8* _1B8;
     u32 _1BC;
     u16 _1C0;
     u16 _1C2;
-    u32* _1C4;
+    u8** _1C4;
     u8** _1C8;
     u16* _1CC;
     u8 _1D0;
--- a/libs/RVL_SDK/include/revolution/gd/GDGeometry.h
+++ b/libs/RVL_SDK/include/revolution/gd/GDGeometry.h
@@ -1,5 +1,9 @@
 #ifndef GDGEOMETRY_H
 #define GDGEOMETRY_H
+
+#include "revolution/gx/GXEnum.h"
+#include "revolution/gx/GXStruct.h"
+#include "revolution/types.h"
 
 #include "revolution/gd/GDBase.h"
 #include "revolution/gx/GXEnum.h"
@@ -190,4 +194,16 @@
 }
 #endif
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void GDSetCullMode(GXCullMode);
+void GDSetGenMode2(u8, u8, u8, u8, GXCullMode);
+void GDSetTexCoordGen(GXTexCoordID, GXTexGenType, GXTexGenSrc, u8, u32);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif  // GDGEOMETRY_H
--- /dev/null
+++ b/pc-port/src/revolution/gd/GDGeometry.h
@@ -0,0 +1,2 @@
+#pragma once
+#include <dolphin/gd/GDGeometry.h>
--- a/libs/RVL_SDK/include/revolution/gd/GDTev.h
+++ b/libs/RVL_SDK/include/revolution/gd/GDTev.h
@@ -1,5 +1,9 @@
 #ifndef GDTEV_H
 #define GDTEV_H
+
+#include "revolution/gx/GXEnum.h"
+#include "revolution/gx/GXStruct.h"
+#include "revolution/types.h"
 
 #define TEV_COLOR_ENV_0_ID 0x000000c0
 #define TEV_ALPHA_ENV_0_ID 0x000000c1
@@ -163,4 +167,18 @@
     ((u32)(map0) << 0 | (u32)(coord0) << 3 | (u32)(enable0) << 6 | (u32)(color0) << 7 | (u32)(map1) << 12 | (u32)(coord1) << 15 |                    \
      (u32)(enable1) << 18 | (u32)(color1) << 19 | (u32)(id) << 24)
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void GDSetAlphaCompare(GXCompare, u8, GXAlphaOp, GXCompare, u8);
+void GDSetTevAlphaCalcAndSwap(GXTevStageID, GXTevAlphaArg, GXTevAlphaArg, GXTevAlphaArg, GXTevAlphaArg, GXTevOp, GXTevBias, GXTevScale, u8, GXTevRegID, GXTevSwapSel, GXTevSwapSel);
+void GDSetTevOrder(GXTevStageID, GXTexCoordID, GXTexMapID, GXChannelID, GXTexCoordID, GXTexMapID, GXChannelID);
+void GDSetTevColorCalc(GXTevStageID, GXTevColorArg, GXTevColorArg, GXTevColorArg, GXTevColorArg, GXTevOp, GXTevBias, GXTevScale, u8, GXTevRegID);
+void GDSetTevColor(GXTevRegID, GXColor);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif  // GDTEV_H
--- /dev/null
+++ b/pc-port/src/revolution/gd/GDTev.h
@@ -0,0 +1,2 @@
+#pragma once
+#include <dolphin/gd/GDTev.h>
--- a/libs/RVL_SDK/include/revolution/gd/GDPixel.h
+++ b/libs/RVL_SDK/include/revolution/gd/GDPixel.h
@@ -1,5 +1,9 @@
 #ifndef GDPIXEL_H
 #define GDPIXEL_H
+
+#include "revolution/gx/GXEnum.h"
+#include "revolution/gx/GXStruct.h"
+#include "revolution/types.h"
 
 #define TEV_FOG_PARAM_0_ID 0x000000ee
 #define TEV_FOG_PARAM_1_ID 0x000000ef
@@ -112,4 +116,16 @@
 
 #define BP_TOKEN(token, id) ((u32)(token) << 0 | (u32)(id) << 24)
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void GDSetZMode(u8, GXCompare, u8);
+void GDSetBlendModeEtc(GXBlendMode, GXBlendFactor, GXBlendFactor, GXLogicOp, u8, u8, u8);
+void GDSetDstAlpha(u8, u8);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif  // GDPIXEL_H
--- /dev/null
+++ b/pc-port/src/revolution/gd/GDPixel.h
@@ -0,0 +1,2 @@
+#pragma once
+#include <dolphin/gd/GDPixel.h>
--- a/libs/RVL_SDK/include/revolution/gd/GDIndirect.h
+++ b/libs/RVL_SDK/include/revolution/gd/GDIndirect.h
@@ -1,5 +1,9 @@
 #ifndef GDINDIRECT_H
 #define GDINDIRECT_H
+
+#include "revolution/gx/GXEnum.h"
+#include "revolution/gx/GXStruct.h"
+#include "revolution/types.h"
 
 #define IND_CMD0_ID 0x00000010
 
@@ -31,4 +35,14 @@
 
 #define BP_IND_MASK(mask, id) ((u32)(mask) << 0 | (u32)(id) << 24)
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void GDSetTevDirect(GXTevStageID);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif  // GDINDIRECT_H
--- /dev/null
+++ b/pc-port/src/revolution/gd/GDIndirect.h
@@ -0,0 +1,2 @@
+#pragma once
+#include <dolphin/gd/GDIndirect.h>
--- a/libs/RVL_SDK/include/revolution/gd/GDLight.h
+++ b/libs/RVL_SDK/include/revolution/gd/GDLight.h
@@ -1,5 +1,9 @@
 #ifndef GDLIGHT_H
 #define GDLIGHT_H
+
+#include "revolution/gx/GXEnum.h"
+#include "revolution/gx/GXStruct.h"
+#include "revolution/types.h"
 
 #include "revolution/gx/GXEnum.h"
 #include "revolution/types.h"
@@ -60,4 +64,16 @@
 }
 #endif
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void GDSetChanCtrl(GXChannelID, u8, GXColorSrc, GXColorSrc, u32, GXDiffuseFn, GXAttnFn);
+void GDSetChanAmbColor(GXChannelID, GXColor);
+void GDSetChanMatColor(GXChannelID, GXColor);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif  // GDLIGHT_H
--- /dev/null
+++ b/pc-port/src/revolution/gd/GDLight.h
@@ -0,0 +1,2 @@
+#pragma once
+#include <dolphin/gd/GDLight.h>
--- a/libs/RVL_SDK/include/revolution/gd/GDTexture.h
+++ b/libs/RVL_SDK/include/revolution/gd/GDTexture.h
@@ -1,5 +1,9 @@
 #ifndef GDTEXTURE_H
 #define GDTEXTURE_H
+
+#include "revolution/gx/GXEnum.h"
+#include "revolution/gx/GXStruct.h"
+#include "revolution/types.h"
 
 #define TX_SETMODE0_I0_ID 0x80
 #define TX_SETMODE0_I1_ID 0x81
@@ -97,4 +101,16 @@
 
 #define BP_TEX_CACHE_ODD(tmem, size0, size1, id) ((u32)(tmem) << 0 | (u32)(size0) << 15 | (u32)(size1) << 18 | (u32)(id) << 24)
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void GDSetTexLookupMode(GXTexMapID, GXTexWrapMode, GXTexWrapMode, GXTexFilter, GXTexFilter, f32, f32, f32, u8, u8, GXAnisotropy);
+void GDSetTexImgAttr(GXTexMapID, u16, u16, GXTexFmt);
+void GDSetTexImgPtr(GXTexMapID, void*);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif  // GDTEXTURE_H
--- /dev/null
+++ b/pc-port/src/revolution/gd/GDTexture.h
@@ -0,0 +1,2 @@
+#pragma once
+#include <dolphin/gd/GDTexture.h>
