--- a/pc-port/src/Game/Map/PlanetMap.hpp
+++ b/pc-port/src/Game/Map/PlanetMap.hpp
@@ -9,7 +9,7 @@
 
 class PlanetMap : public LiveActor {
 public:
-    PlanetMap(const char*, const char*);
+    PlanetMap(const char*, const char* = nullptr);
 
     virtual ~PlanetMap();
     virtual void init(const JMapInfoIter&);
@@ -53,7 +53,7 @@
 
 class PlanetMapAnimLow : public PlanetMap {
 public:
-    PlanetMapAnimLow();
+    PlanetMapAnimLow(const char* pName) : PlanetMap(pName, nullptr) {}
 
     virtual ~PlanetMapAnimLow();
     virtual s32 getLowMovementType() const;
--- /dev/null
+++ b/pc-port/src/Game/Map/PlanetMapCreator.cpp
@@ -0,0 +1,444 @@
+#include "Game/Map/LavaSunPlanet.hpp"
+#include "Game/Map/PlanetMap.hpp"
+#include "Game/Map/PlanetMapCreator.hpp"
+#include "Game/Map/PlanetMapWithoutHighModel.hpp"
+#include "Game/MapObj/AstroMapObj.hpp"
+#include "Game/MapObj/BeamGoRoundPlanet.hpp"
+#include "Game/MapObj/BumpAppearPlanet.hpp"
+#include "Game/MapObj/FlexibleSphere.hpp"
+#include "Game/MapObj/HatchWaterPlanet.hpp"
+#include "Game/MapObj/LavaJamboSunPlanet.hpp"
+#include "Game/MapObj/MarblePlanet.hpp"
+#include "Game/MapObj/PeachCastleGardenPlanet.hpp"
+#include "Game/MapObj/QuestionBoxGalleryObj.hpp"
+#include "Game/MapObj/ReverseGravityRoomPlanet.hpp"
+#include "Game/MapObj/RotateMoveObj.hpp"
+#include "Game/MapObj/SandCapsuleInsidePlanet.hpp"
+#include "Game/MapObj/SimpleClipPartsObj.hpp"
+#include "Game/MapObj/SimpleMapObj.hpp"
+#include "Game/MapObj/SnowCapsulePlanet.hpp"
+#include "Game/MapObj/WormEatenPlanet.hpp"
+#include "Game/NameObj/NameObjArchiveListCollector.hpp"
+#include "Game/NameObj/NameObjFactory.hpp"
+#include "Game/Scene/SceneObjHolder.hpp"
+#include "Game/Util/ModelUtil.hpp"
+#include "Game/Util/ObjUtil.hpp"
+#include "Game/Util/SceneUtil.hpp"
+#include "Game/Util/StringUtil.hpp"
+#include <cstdio>
+#include <cstring>
+
+class PlanetMapFarClippable : public PlanetMap {
+public:
+    PlanetMapFarClippable(const char* pName) : PlanetMap(pName, nullptr) {}
+
+    virtual ~PlanetMapFarClippable() {}
+    virtual f32 getFarClipDistance() const { return 50.0f; }
+};
+
+namespace {
+    static const PlanetEntry sUniquePlanetCreateFuncTable[] = {
+        {
+            "BeamGoRoundPlanet",
+            createNameObj< BeamGoRoundPlanet >,
+        },
+        {
+            "BumpAppearPlanet",
+            createNameObj< BumpAppearPlanet >,
+        },
+        {
+            "ChoConveyorPlanetB",
+            createNameObj< RailPlanetMap >,
+        },
+        {
+            "ChoConveyorPlanetD",
+            createNameObj< RotateMoveObj >,
+        },
+        {
+            "DinoPackunBattlePlanet",
+            createNameObj< FurPlanetMap >,
+        },
+        {
+            "DarkHopperPlanetA",
+            createNameObj< SimpleClipPartsObj >,
+        },
+        {
+            "DarkHopperPlanetB",
+            createNameObj< SimpleClipPartsObj >,
+        },
+        {
+            "DarkHopperPlanetC",
+            createNameObj< SimpleClipPartsObj >,
+        },
+        {
+            "DarkHopperPlanetD",
+            createNameObj< SimpleClipPartsObj >,
+        },
+        {
+            "DarknessRoomPlanet",
+            createNameObj< SimpleClipPartsObj >,
+        },
+        {
+            "FlagDiscPlanetB",
+            createNameObj< PlanetMapFarClippable >,
+        },
+        {
+            "FlagDiscPlanetC",
+            createNameObj< PlanetMapFarClippable >,
+        },
+        {
+            "FlagDiscPlanetD",
+            createNameObj< RotateMoveObj >,
+        },
+        {
+            "FlagDiscPlanetE",
+            createNameObj< PlanetMapFarClippable >,
+        },
+        {
+            "HatchWaterPlanet",
+            createNameObj< HatchWaterPlanet >,
+        },
+        {
+            "HeavensDoorInsidePlanet",
+            createNameObj< SimpleMapObj >,
+        },
+        {
+            "HoneyQueenPlanet",
+            createNameObj< FurPlanetMap >,
+        },
+        {
+            "LavaJamboSunPlanet",
+            createNameObj< LavaJamboSunPlanet >,
+        },
+        {
+            "LavaRotatePlanet",
+            createNameObj< LavaSunPlanet >,
+        },
+        {
+            "LavaDomedPlanet",
+            createNameObj< LavaSunPlanet >,
+        },
+        {
+            "MarblePlanet",
+            createNameObj< MarblePlanet >,
+        },
+        {
+            "PeachCastleGardenPlanet",
+            createNameObj< PeachCastleGardenPlanet >,
+        },
+        {
+            "SandCapsuleInsidePlanet",
+            createNameObj< SandCapsuleInsidePlanet >,
+        },
+        {
+            "ScaleDownRelayPlanet",
+            createNameObj< FlexibleSphere >,
+        },
+        {
+            "SnowCapsulePlanet",
+            createNameObj< SnowCapsulePlanet >,
+        },
+        {
+            "TeresaRoomPlanet",
+            createNameObj< PlanetMapAnimLow >,
+        },
+        {
+            "TridentPlanet",
+            createNameObj< AstroSimpleObj >,
+        },
+        {
+            "QuestionBoxPlanetA",
+            createNameObj< QuestionBoxGalleryObj >,
+        },
+        {
+            "QuestionBoxPlanetB",
+            createNameObj< QuestionBoxGalleryObj >,
+        },
+        {
+            "QuestionBoxPlanetC",
+            createNameObj< QuestionBoxGalleryObj >,
+        },
+        {
+            "QuestionBoxPlanetD",
+            createNameObj< QuestionBoxGalleryObj >,
+        },
+        {
+            "QuestionBoxPlanetE",
+            createNameObj< QuestionBoxGalleryObj >,
+        },
+        {
+            "Quicksand2DPlanet",
+            createNameObj< RailPlanetMap >,
+        },
+        {
+            "ReverseGravityRoomPlanet",
+            createNameObj< ReverseGravityRoomPlanet >,
+        },
+        {
+            "SandStreamHighTowerPlanet",
+            createNameObj< RailPlanetMap >,
+        },
+        {
+            "SandStreamJointPlanetA",
+            createNameObj< RailPlanetMap >,
+        },
+        {
+            "SandStreamJointPlanetB",
+            createNameObj< RailPlanetMap >,
+        },
+        {
+            "StarDustStartPlanet",
+            createNameObj< RotateMoveObj >,
+        },
+        {
+            "WormEatenPlanet",
+            createNameObj< WormEatenPlanet >,
+        },
+    };
+    inline const PlanetEntry* findUniquePlanet(const char* pName) {
+        for (u32 i = 0; i < ARRAY_SIZE(sUniquePlanetCreateFuncTable); i++) {
+            const PlanetEntry* pEntry = &sUniquePlanetCreateFuncTable[i];
+            if (MR::isEqualString(pName, pEntry->mName)) {
+                return pEntry;
+            }
+        }
+
+        return nullptr;
+    }
+
+    static const UniqueEntry sUniquePlanetUniqueArchiveName[] = {
+        {
+            "BeamGoRoundPlanet",
+            "BeamGoRoundBeam",
+        },
+        {
+            "BeamGoRoundPlanet",
+            "BeamGoRoundBeamVolume",
+        },
+        {
+            "BeamGoRoundPlanet",
+            "BeamGoRoundBeamBloom",
+        },
+        {
+            "MarblePlanet",
+            "MarblePlanetCore",
+        },
+        {
+            "MarblePlanet",
+            "MarblePlanetElectron",
+        },
+        {
+            "MarblePlanet",
+            "MarblePlanetElectronShadow",
+        },
+        {
+            "OceanRingPlanet",
+            "OceanRingPlanetLowInWater",
+        },
+        {
+            "WormEatenPlanet",
+            "GreenCaterpillarBigFace",
+        },
+        {
+            "WormEatenPlanet",
+            "GreenCaterpillarBigFaceLow",
+        },
+        {
+            "WormEatenPlanet",
+            "GreenCaterpillarBigBody",
+        },
+        {
+            "WormEatenPlanet",
+            "GreenCaterpillarBigBodyLow",
+        },
+        {
+            "WormEatenPlanet",
+            "WormEatenHill",
+        },
+        {
+            "WormEatenPlanet",
+            "WormEatenPlanetLow",
+        },
+    };
+    static const char* sArcName = "PlanetMapDataTable.arc";
+    static const char* sFileName = "PlanetMapDataTable.bcsv";
+
+    void makeSubModelName(const char** pOut, const JMapInfo* pInfo, s32 index, const char* pFlag, const char* pSuffix) {
+        s32 flag;
+        const char* planetName;
+        MR::getCsvDataS32(&flag, pInfo, pFlag, index);
+        MR::getCsvDataStr(&planetName, pInfo, "PlanetName", index);
+
+        if (flag) {
+            u32 size = strlen(planetName) + strlen(pSuffix) + 1;
+            char* name = new char[size];
+            snprintf(name, size, "%s%s", planetName, pSuffix);
+            *pOut = name;
+        } else {
+            *pOut = nullptr;
+        }
+    }
+
+    static bool isDataForceLow(const PlanetMapData* pMapData) {
+        for (int i = 0; i < 8; i++) {
+            if (!MR::isEqualString(pMapData->mForceScenarioData[i], "")) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+};  // namespace
+
+// FIXME: Scheduling issues
+PlanetMapCreator::PlanetMapCreator(const char* pName) : NameObj(pName), mPlanetMapData(nullptr) {
+    JMapInfo* pIter = MR::createCsvParser(::sArcName, ::sFileName);
+
+    mTableCount = MR::getCsvDataElementNum(pIter);
+    mPlanetMapData = new PlanetMapData*[mTableCount];
+    createPlanetMapDataTable(pIter);
+}
+
+CreatorFuncPtr PlanetMapCreator::getCreateFunc(const char* pParam1) {
+    if (isScenarioForceLow(getTableData(pParam1))) {
+        return createNameObj< PlanetMapWithoutHighModel >;
+    }
+
+    const PlanetEntry* pEntry = ::findUniquePlanet(pParam1);
+
+    if (pEntry != nullptr) {
+        return pEntry->mCreateFunc;
+    }
+
+    return createNameObj< PlanetMap >;
+}
+
+void PlanetMapCreator::makeArchiveListPlanet(NameObjArchiveListCollector* pArchiveList, const JMapInfoIter& rIter, const char* pName) {
+    PlanetMapData* pTableData = getTableData(pName);
+    if (isScenarioForceLow(pTableData)) {
+        pArchiveList->addArchive(pTableData->mData[0]);
+    } else {
+        pArchiveList->addArchive(pTableData->mPlanetName);
+
+        const char* pData0 = pTableData->mData[0];
+        if (pData0 != nullptr) {
+            pArchiveList->addArchive(pData0);
+        }
+
+        const char* pData1 = pTableData->mData[1];
+        if (pData1 != nullptr) {
+            pArchiveList->addArchive(pData1);
+        }
+
+        const char* pData2 = pTableData->mData[2];
+        if (pData2 != nullptr) {
+            pArchiveList->addArchive(pData2);
+        }
+
+        const char* pData3 = pTableData->mData[3];
+        if (pData3 != nullptr) {
+            pArchiveList->addArchive(pData3);
+        }
+
+        const char* pData4 = pTableData->mData[4];
+        if (pData4 != nullptr) {
+            pArchiveList->addArchive(pData4);
+        }
+
+        const char* pPlanetName = pTableData->mPlanetName;
+        for (u32 i = 0; i < ARRAY_SIZE(::sUniquePlanetUniqueArchiveName); i++) {
+            const UniqueEntry* pUniqueEntry = &::sUniquePlanetUniqueArchiveName[i];
+            if (MR::isEqualString(pPlanetName, pUniqueEntry->_0)) {
+                pArchiveList->addArchive(pUniqueEntry->_4);
+            }
+        }
+    }
+}
+
+void PlanetMapCreator::createPlanetMapDataTable(JMapInfo* pInfo) {
+    for (int i = 0; i < mTableCount; i++) {
+        mPlanetMapData[i] = new PlanetMapData();
+
+        addTableData(pInfo, i);
+    }
+}
+
+void PlanetMapCreator::addTableData(const JMapInfo* pInfo, s32 idx) {
+    PlanetMapData* curData = mPlanetMapData[idx];
+    MR::getCsvDataStr(&curData->mPlanetName, pInfo, "PlanetName", idx);
+    ::makeSubModelName(&curData->mData[0], pInfo, idx, "LowFlag", "Low");
+    ::makeSubModelName(&curData->mData[1], pInfo, idx, "MiddleFlag", "Middle");
+    ::makeSubModelName(&curData->mData[2], pInfo, idx, "BloomFlag", "Bloom");
+    ::makeSubModelName(&curData->mData[3], pInfo, idx, "WaterFlag", "Water");
+    ::makeSubModelName(&curData->mData[4], pInfo, idx, "IndirectFlag", "Indirect");
+
+    for (int i = 0; i < 5; i++) {
+        if (curData->mData[i]) {
+            if (!MR::isExistModel(curData->mData[i])) {
+                curData->mData[i] = 0;
+            }
+        }
+    }
+
+    for (int i = 0; i < 8; i++) {
+        char data[32];
+        snprintf(data, sizeof(data), "ForceLowScenarioName%d", i);
+
+        MR::getCsvDataStr(&curData->mForceScenarioData[i], pInfo, data, idx);
+    }
+}
+
+PlanetMapData* PlanetMapCreator::getTableData(const char* pName) const {
+    for (int i = 0; i < mTableCount; i++) {
+        if (MR::isEqualString(pName, mPlanetMapData[i]->mPlanetName)) {
+            return mPlanetMapData[i];
+        }
+    }
+
+    return nullptr;
+}
+
+bool PlanetMapCreator::isScenarioForceLow(const PlanetMapData* pData) const {
+    bool isForce = ::isDataForceLow(pData);
+
+    if (!isForce) {
+        return false;
+    }
+
+    char name[64];
+    snprintf(name, sizeof(name), "%s_%d", MR::getCurrentStageName(), MR::getCurrentScenarioNo());
+
+    for (int i = 0; i < 8; i++) {
+        if (MR::isEqualString(pData->mForceScenarioData[i], "")) {
+            return false;
+        }
+
+        if (MR::isEqualString(name, pData->mForceScenarioData[i])) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void PlanetMapCreatorFunction::makeArchiveList(NameObjArchiveListCollector* pArchiveList, const JMapInfoIter& rIter, const char* pName) {
+    MR::getSceneObj< PlanetMapCreator >(SceneObj_PlanetMapCreator)->makeArchiveListPlanet(pArchiveList, rIter, pName);
+}
+
+CreatorFuncPtr PlanetMapCreatorFunction::getPlanetMapCreator(const char* pName) {
+    return MR::getSceneObj< PlanetMapCreator >(SceneObj_PlanetMapCreator)->getCreateFunc(pName);
+}
+
+bool PlanetMapCreatorFunction::isLoadArchiveAfterScenarioSelected(const char* pArchive) {
+    bool isExistTableData = MR::getSceneObj< PlanetMapCreator >(SceneObj_PlanetMapCreator)->getTableData(pArchive);
+
+    if (!isExistTableData) {
+        return false;
+    }
+
+    return ::isDataForceLow(MR::getSceneObj< PlanetMapCreator >(SceneObj_PlanetMapCreator)->getTableData(pArchive));
+}
+
+bool PlanetMapCreatorFunction::isRegisteredObj(const char* pArchive) {
+    return MR::getSceneObj< PlanetMapCreator >(SceneObj_PlanetMapCreator)->getTableData(pArchive);
+}
--- /dev/null
+++ b/pc-port/src/Game/Map/PlanetMapCreator.hpp
@@ -0,0 +1,48 @@
+#pragma once
+
+#include "Game/NameObj/NameObj.hpp"
+
+class NameObjArchiveListCollector;
+
+struct PlanetMapData {
+    const char* mPlanetName;            // 0x0
+    const char* mData[5];               // 0x4
+    const char* mForceScenarioData[8];  // 0x18
+};
+
+struct UniqueEntry {
+    const char* _0;
+    const char* _4;
+};
+
+typedef NameObj* (*CreatorFuncPtr)(const char*);
+
+struct PlanetEntry {
+    /* 0x0 */ const char* mName;
+    /* 0x4 */ CreatorFuncPtr mCreateFunc;
+};
+
+class PlanetMapCreator : public NameObj {
+public:
+    PlanetMapCreator(const char*);
+
+    CreatorFuncPtr getCreateFunc(const char*);
+    void makeArchiveListPlanet(NameObjArchiveListCollector*, const JMapInfoIter&, const char*);
+    void createPlanetMapDataTable(JMapInfo*);
+    void addTableData(const JMapInfo*, s32);
+    PlanetMapData* getTableData(const char*) const;
+    bool isScenarioForceLow(const PlanetMapData*) const;
+
+    PlanetMapData** mPlanetMapData;  // 0xC
+    s32 mTableCount;                 // 0x10
+};
+
+class PlanetMapCreatorFunction {
+public:
+    static void makeArchiveList(NameObjArchiveListCollector*, const JMapInfoIter&, const char*);
+
+    static bool isLoadArchiveAfterScenarioSelected(const char*);
+    static bool isRegisteredObj(const char*);
+
+    static CreatorFuncPtr getPlanetMapCreator(const char*);
+};
--- a/pc-port/src/Game/Util/ObjUtil.hpp
+++ b/pc-port/src/Game/Util/ObjUtil.hpp
@@ -86,6 +86,7 @@
 
     bool isExistFileInArc(const ResourceHolder*, const char*, ...);
     JMapInfo* createCsvParser(const ResourceHolder*, const char*, ...);
+    JMapInfo* createCsvParser(const char*, const char*, ...);
     JMapInfo* tryCreateCsvParser(const LiveActor*, const char*, ...);
     JMapInfo* tryCreateCsvParser(const ResourceHolder*, const char*, ...);
     s32 getCsvDataElementNum(const JMapInfo*);
