diff --git a/src/JSystem/JKernel/JKRHeap.hpp b/src/JSystem/JKernel/JKRHeap.hpp
index 8fc9225fd..ac0151ea4 100644
--- a/src/JSystem/JKernel/JKRHeap.hpp
+++ b/src/JSystem/JKernel/JKRHeap.hpp
@@ -61,7 +61,7 @@ public:
     virtual bool state_compare(const TState&, const TState&) const;
     virtual void state_dump(const TState&) const;
 
-    void* alloc(u32, int);
+    void* alloc(u32, int) NO_INLINE;
     JKRHeap* becomeSystemHeap();
     JKRHeap* becomeCurrentHeap();
     void destroy() {
@@ -140,7 +140,7 @@ public:
     static bool initArena(char** memory, u32* size, int maxHeaps);
     static bool initArena2(char** memory, u32* size, int maxHeaps);
     static void* alloc(u32 size, int alignment, JKRHeap* heap);
-    static void free(void* ptr, JKRHeap* heap);
+    static void free(void* ptr, JKRHeap* heap) NO_INLINE;
     static s32 resize(void* ptr, u32 size, JKRHeap* heap);
     static s32 getSize(void* ptr, JKRHeap* heap);
     static JKRHeap* findFromRoot(void* ptr);
@@ -201,8 +201,6 @@ public:
     static void* mUserRamEnd;
     static u32 mMemorySize;
 
-    static u32 ARALT_AramStartAddr;
-
     inline void* getStartAddr() const {
         return (void*)mStart;
     }
diff --git a/src/compat/JKRHeapCompat.cpp b/src/compat/JKRHeapCompat.cpp
index 8a4689248..2cf19f838 100644
--- a/src/compat/JKRHeapCompat.cpp
+++ b/src/compat/JKRHeapCompat.cpp
@@ -30,7 +30,7 @@ static bool byte_806B70B8;
 
 u32 JKRHeap::mMemorySize;
 
-u32 JKRHeap::ARALT_AramStartAddr = 0x90000000;
+u32 ARALT_AramStartAdr = 0x90000000;
 
 JKRHeap::JKRHeap(void* data, u32 size, JKRHeap* parent, bool error) : JKRDisposer(), mChildTree(this), mDisposerList() {
     OSInitMutex(&mMutex);
@@ -224,26 +224,25 @@ JKRHeap* JKRHeap::findAllHeap(void* ptr) const {
 }
 
 void JKRHeap::dispose_subroutine(uintptr_t start, uintptr_t end) {
+    JSUListIterator< JKRDisposer > it(mDisposerList.getFirst());
     JSUListIterator< JKRDisposer > last_it;
-    JSUListIterator< JKRDisposer > next_it;
-    JSUListIterator< JKRDisposer > it;
 
-    for (it = mDisposerList.getFirst(); it != mDisposerList.getEnd(); it = next_it) {
-        JKRDisposer* disp = it.getObject();
+    JSULink< JKRDisposer >* link;
+    while ((link = it.mLink) != nullptr) {
+        JKRDisposer* disp = link->getObject();
 
-        if ((void*)start <= disp && disp < (void*)end) {
+        if (reinterpret_cast< void* >(start) <= disp && disp < reinterpret_cast< void* >(end)) {
             disp->~JKRDisposer();
 
             if (last_it == nullptr) {
-                next_it = mDisposerList.getFirst();
+                it = mDisposerList.getFirst();
             } else {
-                next_it = last_it;
-                next_it++;
+                it = last_it;
+                it++;
             }
         } else {
             last_it = it;
-            next_it = it;
-            next_it++;
+            it++;
         }
     }
     smgpc::compat::detail::finalize_jkr_heap_objects(this, start, end);
@@ -314,11 +313,11 @@ void JKRHeap::state_dump(const TState&) const {
 }
 
 void JKRHeap::setAltAramStartAdr(u32 addr) {
-    ARALT_AramStartAddr = addr;
+    ARALT_AramStartAdr = addr;
 }
 
 u32 JKRHeap::getAltAramStartAdr() {
-    return ARALT_AramStartAddr;
+    return ARALT_AramStartAdr;
 }
 
 s32 JKRHeap::do_changeGroupID(u8) {
