--- a/pc-port/src/Game/Player/MarioActorSpecialDraw.cpp
+++ b/pc-port/src/Game/Player/MarioActorSpecialDraw.cpp
@@ -2,6 +2,8 @@
 #include "Game/Player/J3DModelX.hpp"
 #include "Game/Player/MarioActor.hpp"
 #include "Game/Player/MarioConst.hpp"
+#include "Game/Player/MarioState.hpp"
+#include "Game/Screen/FullScreenBlur.hpp"
 #include "Game/Player/MarioParts.hpp"
 #include "Game/LiveActor/HitSensor.hpp"
 #include "Game/Util/DirectDraw.hpp"
@@ -26,6 +28,10 @@
 #include <revolution/gx/GXPixel.h>
 #include <revolution/gx/GXCull.h>
 #include <revolution/gx/GXTransform.h>
+#include <revolution/gx/GXTexture.h>
+#include <revolution/gx/GXGeometry.h>
+#include <revolution/gx/GXManage.h>
+#include <revolution/gx/GXVert.h>
 #include <revolution/os/OSCache.h>
 
 void MarioActor::initScreenBox() {
@@ -34,10 +40,10 @@
     mScreenBoxMin.x = 0.0f;
     mScreenBoxMax.y = 0.0f;
     mScreenBoxMax.x = 0.0f;
-    _B38 = 0.0f;
-    _B34 = 0.0f;
-    _B40 = 0.0f;
-    _B3C = 0.0f;
+    mScreenBoxPosition.y = 0.0f;
+    mScreenBoxPosition.x = 0.0f;
+    mScreenBoxSize.y = 0.0f;
+    mScreenBoxSize.x = 0.0f;
 }
 
 bool MarioActor::isUseScreenBox() const {
@@ -96,34 +102,34 @@
         width = 2;
     if (height == 0)
         height = 2;
-    _B34 = x;
-    _B38 = y;
-    _B3C = width;
-    _B40 = height;
-
-    TBox2f box(_B34, _B38, _B34 + _B3C, _B38 + _B40);
+    mScreenBoxPosition.x = x;
+    mScreenBoxPosition.y = y;
+    mScreenBoxSize.x = width;
+    mScreenBoxSize.y = height;
+
+    TBox2f box(mScreenBoxPosition.x, mScreenBoxPosition.y, mScreenBoxPosition.x + mScreenBoxSize.x, mScreenBoxPosition.y + mScreenBoxSize.y);
     TBox2f screen(0.0f, 0.0f, MR::getFrameBufferWidth(), MR::getScreenHeight());
     TBox2f clipped(box);
     if (!clipped.intersect(screen)) {
         clipped.i.set(0.0f);
         clipped.f.set(0.0f);
     }
-    _B34 = clipped.i.x;
-    _B38 = clipped.i.y;
-    _B3C = clipped.getWidth();
-    _B40 = clipped.getHeight();
+    mScreenBoxPosition.x = clipped.i.x;
+    mScreenBoxPosition.y = clipped.i.y;
+    mScreenBoxSize.x = clipped.getWidth();
+    mScreenBoxSize.y = clipped.getHeight();
 }
 
 void MarioActor::captureScreenBox() const {
     if (!isUseScreenBox()) {
         return;
     }
-    if (_B3C < 1.0f || _B40 < 1.0f) {
+    if (mScreenBoxSize.x < 1.0f || mScreenBoxSize.y < 1.0f) {
         return;
     }
 
-    GXSetTexCopyDst(_B3C, _B40, GX_TF_RGB565, GX_FALSE);
-    GXSetTexCopySrc(_B34, _B38, _B3C, _B40);
+    GXSetTexCopyDst(mScreenBoxSize.x, mScreenBoxSize.y, GX_TF_RGB565, GX_FALSE);
+    GXSetTexCopySrc(mScreenBoxPosition.x, mScreenBoxPosition.y, mScreenBoxSize.x, mScreenBoxSize.y);
     GXSetColorUpdate(GX_FALSE);
     GXSetAlphaUpdate(GX_TRUE);
 
@@ -136,7 +142,44 @@
     GXSetDstAlpha(GX_FALSE, 0);
 }
 
-// void MarioActor::writeBackScreenBox() const {}
+void MarioActor::writeBackScreenBox() const {
+    if (!isUseScreenBox()) {
+        return;
+    }
+    if (mScreenBoxSize.x < 1.0f || mScreenBoxSize.y < 1.0f) {
+        return;
+    }
+    TDDraw::setup(1, 0, 2);
+    GXTexObj texture;
+    GXInitTexObj(&texture, _B44, mScreenBoxSize.x, mScreenBoxSize.y, GX_TF_RGB565, GX_CLAMP, GX_CLAMP, GX_FALSE);
+    GXLoadTexObj(&texture, GX_TEXMAP0);
+    TVec2f minimum, maximum;
+    MR::convertFrameBufferPosToScreenPos(&minimum, mScreenBoxPosition);
+    MR::convertFrameBufferPosToScreenPos(&maximum, mScreenBoxPosition + mScreenBoxSize + TVec2f(0.99f, 0.99f));
+    GXSetBlendMode(GX_BM_BLEND, GX_BL_INVDSTALPHA, GX_BL_DSTALPHA, GX_LO_NOOP);
+    GXSetDstAlpha(GX_TRUE, 0);
+    GXSetAlphaUpdate(GX_TRUE);
+    GXPixModeSync();
+    GXTexModeSync();
+    s16 right = static_cast< s32 >(maximum.x) + 1;
+    right &= ~1;
+    maximum.x = right;
+    GXSetZMode(GX_TRUE, GX_ALWAYS, GX_TRUE);
+    GXBegin(GX_QUADS, GX_VTXFMT0, 4);
+    GXPosition3f32(minimum.x, minimum.y, 0.9999f);
+    GXTexCoord2f32(0.0f, 0.0f);
+    GXPosition3f32(maximum.x, minimum.y, 0.9999f);
+    GXTexCoord2f32(1.0f, 0.0f);
+    GXPosition3f32(maximum.x, maximum.y, 0.9999f);
+    GXTexCoord2f32(1.0f, 1.0f);
+    GXPosition3f32(minimum.x, maximum.y, 0.9999f);
+    GXTexCoord2f32(0.0f, 1.0f);
+    GXEnd();
+    GXSetDstAlpha(GX_FALSE, 0);
+    GXSetAlphaUpdate(GX_FALSE);
+    TDDraw::close();
+}
+
 
 void MarioActor::calc1stPersonView() {
     f32 length = (mCamPos - mPosition).length();
@@ -331,7 +374,56 @@
     }
 }
 
-// void MarioActor::drawScreenBlend() const {}
+void MarioActor::drawScreenBlend() const {
+    if (mMario->_97C && mMario->_97C->getBlurOffset() != 0.0f) {
+        MR::drawFullScreenBlur(mMario->_97C->getBlurOffset());
+    }
+    if (_1C6) {
+        MR::drawFullScreenBlur(_1C8, _1CC, _1D0, _1D1);
+    }
+    if (mBeeWallWalk) {
+        TVec2f low, high, minimum, maximum;
+        TDDraw::project2D(&low, mMario->mShadowPos);
+        high = low + TVec2f(100.0f, 100.0f);
+        low = low + TVec2f(-100.0f, -100.0f);
+        if (low.x >= high.x) {
+            minimum.x = high.x;
+            maximum.x = low.x;
+        } else {
+            minimum.x = low.x;
+            maximum.x = high.x;
+        }
+        if (low.y >= high.y) {
+            minimum.y = high.y;
+            maximum.y = low.y;
+        } else {
+            minimum.y = low.y;
+            maximum.y = high.y;
+        }
+        low = minimum + TVec2f(-10.0f, -10.0f);
+        high = maximum + TVec2f(10.0f, 10.0f);
+        TDDraw::setup(0, 0, 2);
+        GXSetCullMode(GX_CULL_NONE);
+        GXSetAlphaUpdate(GX_FALSE);
+        GXSetColorUpdate(GX_TRUE);
+        GXSetBlendMode(GX_BM_BLEND, GX_BL_ONE, GX_BL_INVDSTALPHA, GX_LO_NOOP);
+        GXSetDstAlpha(GX_TRUE, 0);
+        TDDraw::drawFillBox(low, high, 0);
+        GXSetAlphaUpdate(GX_FALSE);
+        MR::loadProjectionMtx();
+        MR::loadViewMtx();
+    }
+    if (_BC4) {
+        u32 color = 0xFF000080;
+        if (mMario->_1C._3) {
+            color = 0x00004020;
+        }
+        TDDraw::setup(0, 1, 2);
+        TDDraw::drawFillBox(TVec3f(0.0f, 0.0f, 0.0f), TVec3f(static_cast< f32 >(MR::getScreenWidth()), static_cast< f32 >(MR::getScreenHeight()), 0.0f), color);
+        TDDraw::close();
+    }
+}
+
 
 void MarioActor::updateRandomTexture(f32 value) {
     _B88 = 1 - _B88;
--- a/pc-port/src/Game/Player/MarioActor.hpp
+++ b/pc-port/src/Game/Player/MarioActor.hpp
@@ -827,10 +827,8 @@
     /* 0xB18 */ TVec3f _B18;
     /* 0xB24 */ TVec2f mScreenBoxMin;
     /* 0xB2C */ TVec2f mScreenBoxMax;
-    /* 0xB34 */ f32 _B34;
-    /* 0xB38 */ f32 _B38;
-    /* 0xB3C */ f32 _B3C;
-    /* 0xB40 */ f32 _B40;
+    /* 0xB34 */ TVec2f mScreenBoxPosition;
+    /* 0xB3C */ TVec2f mScreenBoxSize;
     /* 0xB44 */ void* _B44;
     /* 0xB48 */ FootPrint* _B48;
     /* 0xB4C */ IceStep** _B4C;
--- a/pc-port/src/Game/Player/MarioActorInit.cpp
+++ b/pc-port/src/Game/Player/MarioActorInit.cpp
@@ -278,10 +278,10 @@
     mScreenBoxMin.x = 0.0f;
     mScreenBoxMax.y = 0.0f;
     mScreenBoxMax.x = 0.0f;
-    _B38 = 0.0f;
-    _B34 = 0.0f;
-    _B40 = 0.0f;
-    _B3C = 0.0f;
+    mScreenBoxPosition.y = 0.0f;
+    mScreenBoxPosition.x = 0.0f;
+    mScreenBoxSize.y = 0.0f;
+    mScreenBoxSize.x = 0.0f;
     _B44 = 0;
     _B48 = nullptr;
     _B4C = 0;
--- a/pc-port/src/revolution/gx/GXTexture.h
+++ b/pc-port/src/revolution/gx/GXTexture.h
@@ -0,0 +1,2 @@
+#pragma once
+#include <dolphin/gx/GXTexture.h>
--- a/pc-port/src/revolution/gx/GXGeometry.h
+++ b/pc-port/src/revolution/gx/GXGeometry.h
@@ -0,0 +1,2 @@
+#pragma once
+#include <dolphin/gx/GXGeometry.h>
--- a/pc-port/src/revolution/gx/GXManage.h
+++ b/pc-port/src/revolution/gx/GXManage.h
@@ -0,0 +1,2 @@
+#pragma once
+#include <dolphin/gx/GXManage.h>
