--- a/pc-port/src/Game/Map/CollisionCategorizedKeeper.cpp
+++ b/pc-port/src/Game/Map/CollisionCategorizedKeeper.cpp
@@ -347,6 +347,88 @@
     }
 
     _10 = count;
+    return count;
+}
+
+u32 CollisionCategorizedKeeper::createAreaPolygonList(Triangle* pTriangles, u32 capacity, const TVec3f& rStart, const TVec3f& rEnd) {
+    MR::getCollisionDirector();
+    u32 count = 0;
+    TVec3f minimum;
+    TVec3f maximum;
+    if (rStart.x < rEnd.x) {
+        minimum.x = rStart.x;
+        maximum.x = rEnd.x;
+    } else {
+        minimum.x = rEnd.x;
+        maximum.x = rStart.x;
+    }
+    if (rStart.y < rEnd.y) {
+        minimum.y = rStart.y;
+        maximum.y = rEnd.y;
+    } else {
+        minimum.y = rEnd.y;
+        maximum.y = rStart.y;
+    }
+    if (rStart.z < rEnd.z) {
+        minimum.z = rStart.z;
+        maximum.z = rEnd.z;
+    } else {
+        minimum.z = rEnd.z;
+        maximum.z = rStart.z;
+    }
+    for (CollisionZone** pZone = mZones; pZone != mZones + mZoneNum; pZone++) {
+        if (pZone != mZones) {
+            if (!isSphereOverlappingWithBox(minimum, maximum, (*pZone)->_808, (*pZone)->mRadius)) {
+                continue;
+            }
+        }
+        s32 partCount = (*pZone)->mNumParts;
+        for (s32 i = 0; i < partCount; i++) {
+            CollisionParts* pParts = (*pZone)->mPartsArray[i];
+            if (!pParts->_CC) {
+                continue;
+            }
+            f32 radius = pParts->_D8;
+            if (!isSphereOverlappingWithBox(minimum, maximum, pParts->getTrans(), radius)) {
+                continue;
+            }
+            count += pParts->createAreaPolygonList(pTriangles + count, capacity - count, rStart, rEnd);
+            if (capacity <= count) {
+                return count;
+            }
+        }
+    }
+    return count;
+}
+
+u32 CollisionCategorizedKeeper::createAreaPolygonListArray(Triangle* pTriangles, u32 capacity, TVec3f* pPoints, u32 pointCount) {
+    MR::getCollisionDirector();
+    u32 count = 0;
+    TVec3f minimum;
+    TVec3f maximum;
+    MR::createBoundingBox(pPoints, pointCount, &minimum, &maximum);
+    for (CollisionZone** pZone = mZones; pZone != mZones + mZoneNum; pZone++) {
+        if (pZone != mZones) {
+            if (!isSphereOverlappingWithBox(minimum, maximum, (*pZone)->_808, (*pZone)->mRadius)) {
+                continue;
+            }
+        }
+        s32 partCount = (*pZone)->mNumParts;
+        for (s32 i = 0; i < partCount; i++) {
+            CollisionParts* pParts = (*pZone)->mPartsArray[i];
+            if (!pParts->_CC) {
+                continue;
+            }
+            f32 radius = pParts->_D8;
+            if (!isSphereOverlappingWithBox(minimum, maximum, pParts->getTrans(), radius)) {
+                continue;
+            }
+            count += pParts->createAreaPolygonListArray(pTriangles + count, capacity - count, pPoints, pointCount);
+            if (capacity <= count) {
+                return count;
+            }
+        }
+    }
     return count;
 }
 
--- a/pc-port/src/Game/Map/CollisionParts.cpp
+++ b/pc-port/src/Game/Map/CollisionParts.cpp
@@ -498,6 +498,47 @@
     return acceptedCount;
 }
 
+u32 CollisionParts::createAreaPolygonList(Triangle* pTriangles, u32 capacity, const TVec3f& rStart, const TVec3f& rEnd) {
+    KC_PrismData* prisms[512];
+    TPos3f rotation;
+    PSMTXCopy(mInvBaseMatrix.toMtxPtr(), rotation.toMtxPtr());
+    rotation.zeroTrans();
+    TVec3f minimum;
+    TVec3f maximum;
+    mInvBaseMatrix.mult(rStart, minimum);
+    mInvBaseMatrix.mult(rEnd, maximum);
+    u32 count = mServer->checkArea3D(reinterpret_cast< Fxyz* >(&minimum), reinterpret_cast< Fxyz* >(&maximum), prisms, capacity);
+    if (count == 0) {
+        return 0;
+    }
+    for (u32 i = 0; i < count; i++) {
+        pTriangles[i].fillData(this, mServer->toIndex(prisms[i]), mHitSensor);
+    }
+    return count;
+}
+
+u32 CollisionParts::createAreaPolygonListArray(Triangle* pTriangles, u32 capacity, TVec3f* pPoints, u32 pointCount) {
+    KC_PrismData* prisms[512];
+    TVec3f localPoints[32];
+    TPos3f rotation;
+    PSMTXCopy(mInvBaseMatrix.toMtxPtr(), rotation.toMtxPtr());
+    rotation.zeroTrans();
+    for (u32 i = 0; i < pointCount; i++) {
+        mInvBaseMatrix.mult(pPoints[i], localPoints[i]);
+    }
+    TVec3f minimum;
+    TVec3f maximum;
+    MR::createBoundingBox(localPoints, pointCount, &minimum, &maximum);
+    u32 count = mServer->checkArea3D(reinterpret_cast< Fxyz* >(&minimum), reinterpret_cast< Fxyz* >(&maximum), prisms, capacity);
+    if (count == 0) {
+        return 0;
+    }
+    for (u32 i = 0; i < count; i++) {
+        pTriangles[i].fillData(this, mServer->toIndex(prisms[i]), mHitSensor);
+    }
+    return count;
+}
+
 void CollisionParts::calcForceMovePower(TVec3f* a1, const TVec3f& a2) const {
     TVec3f tStack88 = a2;
     TMtx34f auStack76;
--- a/pc-port/src/Game/Map/KCollision.cpp
+++ b/pc-port/src/Game/Map/KCollision.cpp
@@ -140,6 +140,113 @@
         return pPrism;
     }
     return nullptr;
+}
+
+u32 KCollisionServer::checkArea3D(Fxyz* pPointA, Fxyz* pPointB, KC_PrismData** pPrisms, u32 capacity) {
+    u16* pLongestList = nullptr;
+    u16* pPreviousList = nullptr;
+    u32 count = 0;
+    Fxyz points[2];
+    points[0] = *pPointA;
+    points[1] = *pPointB;
+    TVec3f minimum;
+    TVec3f maximum;
+    MR::createBoundingBox(reinterpret_cast< TVec3f* >(points), 2, &minimum, &maximum);
+    if (minimum.x == maximum.x) {
+        minimum.x -= 1.0f;
+        maximum.x += 1.0f;
+    }
+    if (minimum.y == maximum.y) {
+        minimum.y -= 1.0f;
+        maximum.y += 1.0f;
+    }
+    if (minimum.z == maximum.z) {
+        minimum.z -= 1.0f;
+        maximum.z += 1.0f;
+    }
+    V3u localMinimum;
+    V3u localMaximum;
+    if (!outCheck(&minimum, &maximum, &localMinimum, &localMaximum)) {
+        return 0;
+    }
+
+    u32 z = localMinimum.z;
+
+    do {
+        u32 y = localMinimum.y;
+        s32 stepZ = 1000000;
+
+        do {
+            u32 x = localMinimum.x;
+            s32 stepY = 1000000;
+            s32 longestY = 0;
+
+            do {
+                s32 shift;
+                u16* pList = reinterpret_cast< u16* >(searchBlock(&shift, x, y, z));
+                shift = 1 << shift;
+                s32 width = shift;
+                s32 mask = width - 1;
+                s32 remainingZ = width - (z & mask);
+                s32 stepX = width - (x & mask);
+                s32 remainingY = width - (y & mask);
+
+                if (remainingZ < stepZ) {
+                    stepZ = remainingZ;
+                }
+
+                if (remainingY < stepY) {
+                    stepY = remainingY;
+                }
+
+                if (remainingY > longestY && pList[1] != 0) {
+                    longestY = remainingY;
+                    pLongestList = pList;
+                }
+
+                if (pPreviousList == nullptr || pList != pPreviousList) {
+                    while (*++pList != 0) {
+                        KC_PrismData* pPrism = &mFile->mPrisms[*pList];
+
+                        if (pPrism->mHeight <= 0.0f) {
+                            continue;
+                        }
+
+                        KC_PrismData** pEnd = pPrisms + count;
+                        if (std::find(pPrisms, pEnd, static_cast< KC_PrismData* const& >(pPrism)) != pEnd) {
+                            continue;
+                        }
+
+                        TVec3f points[3];
+                        points[0] = getPos(pPrism, 0);
+                        points[1] = getPos(pPrism, 1);
+                        points[2] = getPos(pPrism, 2);
+                        TVec3f prismMinimum;
+                        TVec3f prismMaximum;
+                        MR::createBoundingBox(points, 3, &prismMinimum, &prismMaximum);
+                        if (prismMaximum.x < minimum.x || prismMaximum.y < minimum.y || prismMaximum.z < minimum.z ||
+                            maximum.x < prismMinimum.x || maximum.y < prismMinimum.y || maximum.z < prismMinimum.z) {
+                            continue;
+                        }
+                        *pEnd = pPrism;
+                        count++;
+                        if (count == capacity) {
+                            return capacity;
+                        }
+                    }
+                }
+
+                x += stepX;
+            } while (x <= static_cast< u32 >(localMaximum.x));
+
+            pPreviousList = pLongestList;
+            y += stepY;
+        } while (y <= static_cast< u32 >(localMaximum.y));
+
+        z += stepZ;
+    } while (z <= static_cast< u32 >(localMaximum.z));
+
+    return count;
 }
 
 u32 KCollisionServer::checkSphere(Fxyz* pPos, f32 radius, f32 scale, u32 capacity, KC_PrismData** pPrisms, f32* pDistances, u8* pFeatures) {
