--- a/src/Game/MapObj/EarthenPipe.cpp
+++ b/src/Game/MapObj/EarthenPipe.cpp
@@ -219,7 +219,7 @@
 bool EarthenPipe::isNerveShowUp() const {
     bool ret;
 
-    if (isNerve(&NrvEarthenPipe::EarthenPipeNrvShowUp::sInstance) || isNerve(&NrvEarthenPipe::EarthenPipeNrvShowUp::sInstance)) {
+    if (isNerve(&NrvEarthenPipe::EarthenPipeNrvWaitToShowUp::sInstance) || isNerve(&NrvEarthenPipe::EarthenPipeNrvShowUp::sInstance)) {
         ret = true;
     } else {
         ret = false;
@@ -414,12 +414,10 @@
               !MR::isPlayerDead())) {
             return false;
         }
-        TVec3f sensorPos(TVec3f(mTopJointMtx[0][3], mTopJointMtx[1][3], mTopJointMtx[2][3]));
-        TVec3f playerPos(*MR::getPlayerPos());
-        playerPos.sub(sensorPos);
-        TVec3f delta(playerPos);
+        TVec3f diff = *MR::getPlayerPos() - TVec3f(mTopJointMtx[0][3], mTopJointMtx[1][3], mTopJointMtx[2][3]);
+        TVec3f delta = diff;
         MR::vecKillElement(delta, mGravity, &delta);
-        if (MR::isPlayerSwimming() && PSVECMag(playerPos) > 50.0f && playerPos.dot(_98) < -5.0f) {
+        if (!MR::isPlayerSwimming() && delta.length() > 50.0f && diff.dot(_98) < -5.0f) {
             return false;
         }
         mHostActor = MR::getSensorHost(pSender);
--- a/src/Game/NPC/NPCActor.cpp
+++ b/src/Game/NPC/NPCActor.cpp
@@ -625,8 +625,7 @@
     TVec3f vec2;
     MR::makeAxisFrontUp(&vec1, &vec2, toPlayer, yDir);
     MR::clampVecAngleDeg(&vec2, yDir, f3);
-    // probably a typo
-    return MR::turnQuatYDirRad(&_A0, _A0, vec2, f2 * 0.17453294f) & turned;
+    return MR::turnQuatYDirRad(&_A0, _A0, vec2, f2 * 0.017453292f) & turned;
 }
 
 bool NPCActor::turnToDefault(f32 f1) {
@@ -644,7 +643,7 @@
     f32 dot = zDir.dot(zDir2);
     dot = MR::clamp(dot, -1.0f, 1.0f);
 
-    f32 flt = MR::abs((f1 * 0.17453294f) / MR::acos(dot));
+    f32 flt = MR::abs((f1 * 0.017453292f) / MR::acos(dot));
     flt = MR::clamp(flt, 0.0f, 1.0f);
 
     MR::blendQuatUpFront(&_A0, yDir, zDir, flt, flt);
