--- a/include/Game/Util/HashUtil.hpp
+++ b/include/Game/Util/HashUtil.hpp
@@ -1,23 +1,26 @@
 #pragma once
 
 #include <revolution/types.h>
+#include <stdint.h>
 
 class HashSortTable {
 public:
+    typedef uintptr_t Value;
+
     HashSortTable(u32);
 
-    bool add(const char*, u32, bool);
-    bool add(u32, u32);
-    bool addOrSkip(u32, u32);
+    bool add(const char*, Value, bool);
+    bool add(u32, Value);
+    bool addOrSkip(u32, Value);
     void sort();
-    bool search(u32, u32*);
-    bool search(const char*, u32*);
-    bool search(const char*, const char*, u32*);
+    bool search(u32, Value*);
+    bool search(const char*, Value*);
+    bool search(const char*, const char*, Value*);
     void swap(const char*, const char*);
 
     /* 0x00 */ bool mHasBeenSorted;
     /* 0x04 */ u32* mHashCodes;
-    /* 0x08 */ u32* _8;
+    /* 0x08 */ Value* _8;
     /* 0x0C */ u16* _C;
     /* 0x10 */ u16* _10;
     /* 0x14 */ u32 mCurrentLength;
--- a/include/Game/Player/PlayerEvent.hpp
+++ b/include/Game/Player/PlayerEvent.hpp
@@ -90,7 +90,7 @@
 
     template < typename T >
     void addEventSequence(const char* pName) {
-        mHashTable->add(pName, reinterpret_cast< u32 >(new T()), false);
+        mHashTable->add(pName, reinterpret_cast< HashSortTable::Value >(new T()), false);
     }
 
 private:
--- a/src/Game/Util/HashUtil.cpp
+++ b/src/Game/Util/HashUtil.cpp
@@ -9,7 +9,7 @@
 
 HashSortTable::HashSortTable(u32 cnt) {
     mHashCodes = new u32[cnt];
-    _8 = new u32[cnt];
+    _8 = new Value[cnt];
     _C = new u16[0x100];
     _10 = new u16[0x100];
     mCurrentLength = 0;
@@ -17,7 +17,7 @@
     mHasBeenSorted = false;
 }
 
-bool HashSortTable::add(const char* pName, u32 a2, bool isValidSkip) {
+bool HashSortTable::add(const char* pName, Value a2, bool isValidSkip) {
     u32 hash = MR::getHashCode(pName);
 
     if (isValidSkip) {
@@ -27,7 +27,7 @@
     }
 }
 
-bool HashSortTable::add(u32 a1, u32 a2) {
+bool HashSortTable::add(u32 a1, Value a2) {
     mHashCodes[mCurrentLength] = a1;
     _8[mCurrentLength] = a2;
     mCurrentLength++;
@@ -35,7 +35,7 @@
     return true;
 }
 
-bool HashSortTable::addOrSkip(u32 a1, u32 a2) {
+bool HashSortTable::addOrSkip(u32 a1, Value a2) {
     for (u32 i = 0; i < mCurrentLength; i++) {
         if (a1 == mHashCodes[i]) {
             return false;
@@ -56,7 +56,7 @@
 
     s32 originalIndices[0x400];
     MR::sortSmall(mCurrentLength, mHashCodes, originalIndices);
-    u32 swapArray[0x400];
+    Value swapArray[0x400];
 
     for (int i = 0; i < mCurrentLength; i++) {
         swapArray[i] = _8[originalIndices[i]];
@@ -108,7 +108,7 @@
     mHasBeenSorted = true;
 }
 
-bool HashSortTable::search(u32 a1, u32* a2) {
+bool HashSortTable::search(u32 a1, Value* a2) {
     u8 upperByte = a1 >> 24;
 
     if (a2 != nullptr) {
@@ -142,13 +142,13 @@
     return false;
 }
 
-bool HashSortTable::search(const char* a1, u32* a2) {
+bool HashSortTable::search(const char* a1, Value* a2) {
     u32 hash = MR::getHashCode(a1);
 
     return search(hash, a2);
 }
 
-bool HashSortTable::search(const char* a1, const char* a2, u32* a3) {
+bool HashSortTable::search(const char* a1, const char* a2, Value* a3) {
     u32 hash = MR::getHashCode(a1) + MR::getHashCode(a2);
 
     return search(hash, a3);
--- a/src/Game/Animation/XanimeResource.cpp
+++ b/src/Game/Animation/XanimeResource.cpp
@@ -321,7 +321,7 @@
 
 u32 XanimeResourceTable::getIndexFromHash(u32 hash) const {
     if (mSortTable != nullptr) {
-        u32 index;
+        HashSortTable::Value index;
         if (mSortTable->search(hash, &index)) {
             return index;
         }
@@ -337,7 +337,7 @@
     case 0:
         return -1;
     case 1:
-        u32 position;
+        HashSortTable::Value position;
         if (mSortTable->search(hash, &position)) {
             return position;
         }
--- a/src/Game/LiveActor/EffectKeeper.cpp
+++ b/src/Game/LiveActor/EffectKeeper.cpp
@@ -126,7 +126,7 @@
     }
 
     if (_18 != nullptr && _18->mHasBeenSorted) {
-        u32 result = 0;
+        HashSortTable::Value result = 0;
 
         if (_18->search(pParam1, &result)) {
             return reinterpret_cast< MultiEmitter* >(result);
@@ -293,11 +293,11 @@
         pEmitter->setName(pParam3);
 
         if (_18 != nullptr) {
-            _18->add(pParam3, reinterpret_cast< u32 >(pEmitter), 0);
+            _18->add(pParam3, reinterpret_cast< HashSortTable::Value >(pEmitter), 0);
         }
     } else {
         if (_18 != nullptr) {
-            _18->add(pParam2, reinterpret_cast< u32 >(pEmitter), 0);
+            _18->add(pParam2, reinterpret_cast< HashSortTable::Value >(pEmitter), 0);
         }
     }
 
--- a/src/Game/Player/MarioAnimationEfx.cpp
+++ b/src/Game/Player/MarioAnimationEfx.cpp
@@ -75,7 +75,7 @@
     _10F = 0;
     closeCallback();
 
-    u32 callbackIdx;
+    HashSortTable::Value callbackIdx;
     if (_120->search(pName, &callbackIdx)) {
         _11C = callbackIdx;
 
--- a/src/Game/Player/MarioSound.cpp
+++ b/src/Game/Player/MarioSound.cpp
@@ -1735,7 +1735,7 @@
 }
 
 void Mario::playSoundJ(const char* pSoundName, s32 timing) {
-    u32 index;
+    HashSortTable::Value index;
     if (_96C->search(pSoundName, &index)) {
         switch ((soundlist[index]._8._0 >> 24) & 0x3) {
         case 0:
@@ -1772,7 +1772,7 @@
 }
 
 void Mario::stopSoundJ(const char* pSoundName, u32 delay) {
-    u32 index;
+    HashSortTable::Value index;
     if (_96C->search(pSoundName, &index)) {
         switch ((soundlist[index]._8._0 >> 24) & 0x3) {
         case 0:
--- a/src/Game/Player/MatrixControl.cpp
+++ b/src/Game/Player/MatrixControl.cpp
@@ -22,13 +22,13 @@
 }
 
 u8 MatrixControl::getValue(const char* pName, u8 value) const {
-    u32 index;
+    HashSortTable::Value index;
     _14->search(pName, &index);
     return _C[index]._8[value];
 }
 
 bool MatrixControl::getValueOrNone(const char* pName, u8 idx, u8* value) const {
-    u32 index;
+    HashSortTable::Value index;
     if (!_14->search(pName, &index)) {
         return false;
     }
@@ -41,19 +41,19 @@
 }
 
 bool MatrixControl::getBit(const char* pName, u8 bit) const {
-    u32 index;
+    HashSortTable::Value index;
     _14->search(pName, &index);
     return _C[index]._4 & (1 << (31 - bit));
 }
 
 bool MatrixControl::isExist(const char* pName) const {
-    u32 index;
+    HashSortTable::Value index;
     return _14->search(pName, &index);
 }
 
 bool MatrixControl::getBitOrNone(const char* pName, u8 bit) const {
     // FIXME: some black magickery going on here
-    u32 index;
+    HashSortTable::Value index;
     if (_14->search(pName, &index)) {
         return _1D;
     }
@@ -75,7 +75,7 @@
 }
 
 bool MatrixValueGetter::getValue(const char* pName, f32* value) const {
-    u32 index;
+    HashSortTable::Value index;
     if (!_10->search(pName, &index)) {
         return false;
     }
--- a/src/Game/Player/PlayerEvent.cpp
+++ b/src/Game/Player/PlayerEvent.cpp
@@ -41,8 +41,9 @@
 }
 
 void EventSequencer::startEvent(const char* pName) {
-    EventSequence* sequence;
-    if (mHashTable->search(pName, reinterpret_cast< u32* >(&sequence))) {
+    HashSortTable::Value value;
+    if (mHashTable->search(pName, &value)) {
+        EventSequence* sequence = reinterpret_cast< EventSequence* >(value);
         mSequence = sequence;
         mSequenceFrame = 0;
         sequence->clearFlag();
--- a/src/Game/Player/MarioEffect.cpp
+++ b/src/Game/Player/MarioEffect.cpp
@@ -232,7 +232,7 @@
     _BA4 = new HashSortTable(entryCount);
     entry = cMaterialEffectTable;
     for (int i = 0; i < entryCount; i++) {
-        _BA4->add(cMaterialEffectTable[i].mName, reinterpret_cast< u32 >(&cMaterialEffectTable[i]), false);
+        _BA4->add(cMaterialEffectTable[i].mName, reinterpret_cast< HashSortTable::Value >(&cMaterialEffectTable[i]), false);
     }
 
     _BA4->sort();
@@ -335,7 +335,7 @@
 }
 
 MultiEmitter* MarioActor::playMaterialEffect(const char* pName) {
-    u32 value = 0;
+    HashSortTable::Value value = 0;
 
     _BA4->search(pName, &value);
     MaterialEffectEntry* entry = reinterpret_cast< MaterialEffectEntry* >(value);
@@ -384,7 +384,7 @@
 }
 
 void MarioActor::stopMaterialEffect(const char* pName) {
-    u32 value = 0;
+    HashSortTable::Value value = 0;
 
     _BA4->search(pName, &value);
     MaterialEffectEntry* entry = reinterpret_cast< MaterialEffectEntry* >(value);
