diff --git a/src/Game/Map/PlanetMapCreator.cpp b/src/Game/Map/PlanetMapCreator.cpp
index abf680bdb..26bd25855 100644
--- a/src/Game/Map/PlanetMapCreator.cpp
+++ b/src/Game/Map/PlanetMapCreator.cpp
@@ -229,12 +229,25 @@ namespace {
     static const char* sArcName = "PlanetMapDataTable.arc";
     static const char* sFileName = "PlanetMapDataTable.bcsv";
 
-    void makeSubModelName(const char**, const JMapInfo*, s32, const char*, const char*) {
+    void makeSubModelName(const char** pOut, const JMapInfo* pInfo, s32 index, const char* pFlag, const char* pSuffix) {
+        s32 flag;
+        const char* planetName;
+        MR::getCsvDataS32(&flag, pInfo, pFlag, index);
+        MR::getCsvDataStr(&planetName, pInfo, "PlanetName", index);
+
+        if (flag) {
+            u32 size = strlen(planetName) + strlen(pSuffix) + 1;
+            char* name = new char[size];
+            snprintf(name, size, "%s%s", planetName, pSuffix);
+            *pOut = name;
+        } else {
+            *pOut = nullptr;
+        }
     }
 
     static bool isDataForceLow(const PlanetMapData* pMapData) {
         for (int i = 0; i < 8; i++) {
-            if (!MR::isEqualString(pMapData->mForceScenarioData[i], "Low")) {
+            if (!MR::isEqualString(pMapData->mForceScenarioData[i], "")) {
                 return true;
             }
         }
@@ -370,7 +383,7 @@ bool PlanetMapCreator::isScenarioForceLow(const PlanetMapData* pData) const {
     snprintf(name, sizeof(name), "%s_%d", MR::getCurrentStageName(), MR::getCurrentScenarioNo());
 
     for (int i = 0; i < 8; i++) {
-        if (MR::isEqualString(pData->mForceScenarioData[i], "Low")) {
+        if (MR::isEqualString(pData->mForceScenarioData[i], "")) {
             return false;
         }
 
