diff --git a/include/Game/LiveActor/Binder.hpp b/include/Game/LiveActor/Binder.hpp
index 8cc067cf6..b764d6b26 100644
--- a/include/Game/LiveActor/Binder.hpp
+++ b/include/Game/LiveActor/Binder.hpp
@@ -14,6 +14,11 @@ public:
 
     void setExCollisionParts(CollisionParts* pParts) {
         mExCollisionParts = pParts;
+        if (pParts == nullptr) {
+            _1EC._2 = false;
+        } else {
+            _1EC._2 = true;
+        }
     }
 
     void clear();
diff --git a/src/Game/Map/CollisionCategorizedKeeper.cpp b/src/Game/Map/CollisionCategorizedKeeper.cpp
index a659aee40..7ca42ee91 100644
--- a/src/Game/Map/CollisionCategorizedKeeper.cpp
+++ b/src/Game/Map/CollisionCategorizedKeeper.cpp
@@ -1,4 +1,5 @@
 #include "Game/Map/CollisionCategorizedKeeper.hpp"
+#include "Game/LiveActor/HitSensor.hpp"
 #include "Game/Map/CollisionParts.hpp"
 #include "Game/Map/CollisionDirector.hpp"
 #include "Game/Util/MathUtil.hpp"
@@ -221,3 +222,22 @@ bool CollisionCategorizedKeeper::isSphereOverlappingWithBox(const TVec3f& rMin,
 
     return true;
 }
+
+bool CollisionCategorizedKeeper::searchSameHostParts(CollisionParts** ppParts, CollisionParts* pParts) const {
+    for (CollisionZone* const* zone = mZones; zone != mZones + mZoneNum; zone++) {
+        s32 partCount = (*zone)->mNumParts;
+        for (s32 i = 0; i < partCount; i++) {
+            CollisionParts* part = (*zone)->mPartsArray[i];
+            if (part->mHitSensor->mHost == pParts->mHitSensor->mHost) {
+                *ppParts = part;
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+HitInfo* CollisionCategorizedKeeper::getStrikeInfo(u32 index) {
+    return &mHitInfoArray[index];
+}
diff --git a/src/Game/Util/MapUtil.cpp b/src/Game/Util/MapUtil.cpp
index 92e02692d..900c3cd14 100644
--- a/src/Game/Util/MapUtil.cpp
+++ b/src/Game/Util/MapUtil.cpp
@@ -271,7 +271,29 @@ namespace MR {
         return getCollisionDirector()->getCategoryKeeper(0)->createAreaPolygonListArray(pTriangle, param2, pParam3, param4);
     }
 
-    // trySetMoveLimitCollision
+    bool trySetMoveLimitCollision(LiveActor* pActor) {
+        TVec3f start(pActor->mPosition);
+        TVec3f offset(pActor->mGravity);
+        start -= offset * 150.0f;
+        offset *= 1000.0f;
+
+        if (getCollisionDirector()->getCategoryKeeper(3)->checkStrikeLine(start, offset, 0, nullptr, nullptr) != 0) {
+            HitInfo* pHit = getCollisionDirector()->getCategoryKeeper(3)->getStrikeInfo(0);
+            pActor->mBinder->setExCollisionParts(pHit->mParentTriangle.mParts);
+            return true;
+        }
+
+        if (getCollisionDirector()->getCategoryKeeper(0)->checkStrikeLine(start, offset, 0, nullptr, nullptr) != 0) {
+            HitInfo* pHit = getCollisionDirector()->getCategoryKeeper(0)->getStrikeInfo(0);
+            CollisionParts* pParts = nullptr;
+            CollisionParts* pMapParts = pHit->mParentTriangle.mParts;
+            getCollisionDirector()->getCategoryKeeper(3)->searchSameHostParts(&pParts, pMapParts);
+            pActor->mBinder->setExCollisionParts(pParts);
+            return true;
+        }
+
+        return false;
+    }
 
     bool isBindedGroundIce(const LiveActor* pActor) {
         if (pActor->mBinder == nullptr) {
